{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- author: Macrina Lobo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version '3.10' is out-of-date; the current release version '3.13'\n",
      "  is available with R version '4.1'; see https://bioconductor.org/install\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.10 (BiocManager 1.30.12), R 3.6.3 (2020-02-29)\n",
      "\n",
      "Installing package(s) 'seqLogo'\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//RtmpL5sDHN/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'ade4', 'ape', 'beeswarm', 'BiocManager', 'boot', 'brio',\n",
      "  'broom', 'callr', 'car', 'circlize', 'class', 'cli', 'cluster', 'colorspace',\n",
      "  'corrplot', 'cpp11', 'curl', 'dendextend', 'DEoptimR', 'devtools', 'dplyr',\n",
      "  'e1071', 'fansi', 'fitdistrplus', 'formatR', 'gam', 'gert', 'GGally',\n",
      "  'ggplot2', 'ggsignif', 'gh', 'gtools', 'haven', 'highr', 'htmlTable',\n",
      "  'IRkernel', 'KernSmooth', 'knitr', 'ks', 'lattice', 'leiden', 'lme4', 'MASS',\n",
      "  'Matrix', 'matrixStats', 'mgcv', 'mime', 'multicool', 'mvtnorm', 'nnet',\n",
      "  'openxlsx', 'parallelly', 'plot3D', 'plotly', 'processx', 'proxy',\n",
      "  'quantreg', 'Rcpp', 'RcppArmadillo', 'remotes', 'reticulate',\n",
      "  'rhandsontable', 'rio', 'rmarkdown', 'robustbase', 'rversions', 'segmented',\n",
      "  'seqinr', 'seriation', 'spatial', 'spatstat', 'spatstat.core',\n",
      "  'spatstat.geom', 'spatstat.linnet', 'spatstat.utils', 'statmod', 'survival',\n",
      "  'testthat', 'tidyverse', 'tinytex', 'xfun', 'zip'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(\"seqLogo\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: grid\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(seqLogo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "To cite package ‘seqLogo’ in publications use:\n",
       "\n",
       "  Oliver Bembom (2019). seqLogo: Sequence logos for DNA sequence\n",
       "  alignments. R package version 1.52.0.\n",
       "\n",
       "A BibTeX entry for LaTeX users is\n",
       "\n",
       "  @Manual{,\n",
       "    title = {seqLogo: Sequence logos for DNA sequence alignments},\n",
       "    author = {Oliver Bembom},\n",
       "    year = {2019},\n",
       "    note = {R package version 1.52.0},\n",
       "  }\n",
       "\n",
       "ATTENTION: This citation information has been auto-generated from the\n",
       "package DESCRIPTION file and may need manual editing, see\n",
       "‘help(\"citation\")’.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "citation(\"seqLogo\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//RtmpcWTlq9/downloaded_packages\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"conflicted\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(conflicted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# help(seqLogo.orig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "Loading required package: data.table\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "Loading required package: Matrix\n",
      "\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "Loading required package: rhdf5\n",
      "\n",
      "Loading required package: magrittr\n",
      "\n",
      "Setting default number of Parallel threads to 8.\n",
      "\n",
      "Setting default genome to Mm10.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "\n",
    "#set threads to 1/2 of available cores (assuming each core has 2 threads)\n",
    "addArchRThreads(threads = 8)\n",
    "\n",
    "#set genome\n",
    "addArchRGenome(\"mm10\")\n",
    "\n",
    "setwd(\"/Users/LoboM/Documents/e11_12_analysis/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Successfully loaded ArchRProject!\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "projFoxn1ko = loadArchRProject( \"E11_E12_cisbp2_deviations_distal_motifs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GeneIntegrationMatrix_e11_12_fil_clean'</li><li>'GeneIntegrationMatrix_seRNA_11_12_fil'</li><li>'GeneIntegrationMatrix_unfil'</li><li>'GeneScoreMatrix'</li><li>'Motif2Matrix'</li><li>'MotifCisbp2Matrix'</li><li>'MotifCisbp2_onlyMatrix'</li><li>'MotifCisbp2_only_distalMatrix'</li><li>'MotifCisbp2_only_promoterMatrix'</li><li>'MotifMatrix'</li><li>'MotifWeakMatrix'</li><li>'PeakMatrix'</li><li>'TileMatrix'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'GeneIntegrationMatrix\\_e11\\_12\\_fil\\_clean'\n",
       "\\item 'GeneIntegrationMatrix\\_seRNA\\_11\\_12\\_fil'\n",
       "\\item 'GeneIntegrationMatrix\\_unfil'\n",
       "\\item 'GeneScoreMatrix'\n",
       "\\item 'Motif2Matrix'\n",
       "\\item 'MotifCisbp2Matrix'\n",
       "\\item 'MotifCisbp2\\_onlyMatrix'\n",
       "\\item 'MotifCisbp2\\_only\\_distalMatrix'\n",
       "\\item 'MotifCisbp2\\_only\\_promoterMatrix'\n",
       "\\item 'MotifMatrix'\n",
       "\\item 'MotifWeakMatrix'\n",
       "\\item 'PeakMatrix'\n",
       "\\item 'TileMatrix'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'GeneIntegrationMatrix_e11_12_fil_clean'\n",
       "2. 'GeneIntegrationMatrix_seRNA_11_12_fil'\n",
       "3. 'GeneIntegrationMatrix_unfil'\n",
       "4. 'GeneScoreMatrix'\n",
       "5. 'Motif2Matrix'\n",
       "6. 'MotifCisbp2Matrix'\n",
       "7. 'MotifCisbp2_onlyMatrix'\n",
       "8. 'MotifCisbp2_only_distalMatrix'\n",
       "9. 'MotifCisbp2_only_promoterMatrix'\n",
       "10. 'MotifMatrix'\n",
       "11. 'MotifWeakMatrix'\n",
       "12. 'PeakMatrix'\n",
       "13. 'TileMatrix'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"GeneIntegrationMatrix_e11_12_fil_clean\"\n",
       " [2] \"GeneIntegrationMatrix_seRNA_11_12_fil\" \n",
       " [3] \"GeneIntegrationMatrix_unfil\"           \n",
       " [4] \"GeneScoreMatrix\"                       \n",
       " [5] \"Motif2Matrix\"                          \n",
       " [6] \"MotifCisbp2Matrix\"                     \n",
       " [7] \"MotifCisbp2_onlyMatrix\"                \n",
       " [8] \"MotifCisbp2_only_distalMatrix\"         \n",
       " [9] \"MotifCisbp2_only_promoterMatrix\"       \n",
       "[10] \"MotifMatrix\"                           \n",
       "[11] \"MotifWeakMatrix\"                       \n",
       "[12] \"PeakMatrix\"                            \n",
       "[13] \"TileMatrix\"                            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "getAvailableMatrices(projFoxn1ko)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "motifs = getPeakAnnotation(ArchRProj = projFoxn1ko,name='MotifCisbp2_only_distal')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PWMatrixList of length 1407\n",
       "names(1407): Tfap2a Tfap2a_2 Tfap2a_3 Tfap2a_4 Tfap2a_5 ... Cbfb Nfyc Spz1 Nrf1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "motifs$motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>1</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 1\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 1\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1 1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "PWMatrixToProbMatrix <- function(x){\n",
    "if (class(x) != \"PWMatrix\") stop(\"x must be a TFBSTools::PWMatrix object\")\n",
    "m <- (exp(as(x, \"matrix\"))) * TFBSTools::bg(x)/sum(TFBSTools::bg(x))\n",
    "m <- t(t(m)/colSums(m))\n",
    "m\n",
    "}\n",
    "\n",
    "ProbMatrices <- lapply(motifs$motifs, PWMatrixToProbMatrix)\n",
    "lapply(ProbMatrices, colSums) %>% range\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from scATAC_12_Fig3c_part2\n",
    "to_plot = as.vector(read.csv(\"/Users/LoboM/Dropbox/pharynx_figures_scATAC/JULY_fill_me_in/gene_motif.csv\")[['motif']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "for (entry in to_plot){\n",
    "    filename=paste(\"/Users/LoboM/Dropbox/pharynx_figures_scATAC/JULY_fill_me_in/logos/\",paste(entry,\"seqlogo.eps\",sep=\"_\"),sep='')\n",
    "    # jpeg(filename)\n",
    "    postscript(filename)\n",
    "    seqLogo2(ProbMatrices[[entry]],xaxis=FALSE,yaxis=FALSE,ic.scale=FALSE, fill = c(A=\"#4daf4a\", C=\"#377eb8\", G=\"#ffd92f\", T=\"#e41a1c\"))\n",
    "    \n",
    "    # dev.copy(p,filename)\n",
    "    dev.off()\n",
    "\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "# wanted to customize color so adapted code\n",
    "letterA <- function(x.pos, y.pos, ht, wt, fill = \"#61D04F\", id = NULL) {\n",
    "    x <- c(0, 4, 6, 2, 0, 4, 6, 10, 8, 4, 3.2, 6.8, 6.4, 3.6, 3.2)\n",
    "    y <- c(0, 10, 10, 0, 0, 10, 10, 0, 0, 10, 3, 3, 4, 4, 3)\n",
    "    x <- 0.1 * x\n",
    "    y <- 0.1 * y\n",
    "\n",
    "    x <- .rescale(x)\n",
    "    x <- x.pos + wt * x\n",
    "    y <- y.pos + ht * y\n",
    "\n",
    "    if (is.null(id)) {\n",
    "        id <- c(rep(1L, 5), rep(2L, 5), rep(3L, 5))\n",
    "    } else {\n",
    "        id <- c(rep(id, 5), rep(id + 1L, 5), rep(id + 2L, 5))\n",
    "    }\n",
    "\n",
    "    fill <- rep(fill, 3) # green\n",
    "\n",
    "    list(x = x, y = y, id = id, fill = fill)\n",
    "}\n",
    "\n",
    "#' Draw T letter\n",
    "#'\n",
    "#' @param x.pos numeric. X position for the letter.\n",
    "#' @param y.pos numeric. Y position for the letter.\n",
    "#' @param ht numeric. Height of the letter.\n",
    "#' @param wt numeric. Width of the letter.\n",
    "#' @param fill character. Color of the letter (polygon shape fill)\n",
    "#' @param id integer. Shape id.\n",
    "#'\n",
    "#' @return list. List with 'x', 'y' coordinates, 'id' and 'fill' of the polygon\n",
    "#' shape (letter).\n",
    "#'\n",
    "#' @keywords internal\n",
    "#' @noRd\n",
    "letterT <- function(x.pos, y.pos, ht, wt, fill = \"#DF536B\", id = NULL) {\n",
    "    x <- c(0, 10, 10, 6, 6, 4, 4, 0)\n",
    "    y <- c(10, 10, 9, 9, 0, 0, 9, 9)\n",
    "    x <- 0.1 * x\n",
    "    y <- 0.1 * y\n",
    "\n",
    "    x <- .rescale(x)\n",
    "    x <- x.pos + wt * x\n",
    "    y <- y.pos + ht * y\n",
    "\n",
    "    if (is.null(id)) {\n",
    "        id <- rep(1, 8)\n",
    "    } else {\n",
    "        id <- rep(id, 8)\n",
    "    }\n",
    "\n",
    "    fill <- rep(fill, 1) # red\n",
    "\n",
    "    list(x = x, y = y, id = id, fill = fill)\n",
    "}\n",
    "\n",
    "\n",
    "#' Draw U letter\n",
    "#'\n",
    "#' @param x.pos numeric. X position for the letter.\n",
    "#' @param y.pos numeric. Y position for the letter.\n",
    "#' @param ht numeric. Height of the letter.\n",
    "#' @param wt numeric. Width of the letter.\n",
    "#' @param fill character. Color of the letter (polygon shape fill)\n",
    "#' @param id integer. Shape id.\n",
    "#'\n",
    "#' @return list. List with 'x', 'y' coordinates, 'id' and 'fill' of the polygon\n",
    "#' shape (letter).\n",
    "#'\n",
    "#' @keywords internal\n",
    "#' @noRd\n",
    "letterU <- function(x.pos, y.pos, ht, wt, fill = \"#DF536B\", id = NULL) {\n",
    "    angle1 <- seq(pi / 2, pi, length = 100)\n",
    "    angle2 <- seq(pi, 1.5 * pi, length = 100)\n",
    "    x.l1 <- 0.5 + 0.5 * sin(angle1)\n",
    "    y.l1 <- 0.5 + 0.5 * cos(angle1)\n",
    "    x.l2 <- 0.5 + 0.5 * sin(angle2)\n",
    "    y.l2 <- 0.5 + 0.5 * cos(angle2)\n",
    "\n",
    "    x.l <- c(x.l1, x.l2)\n",
    "    y.l <- c(y.l1, y.l2)\n",
    "\n",
    "    x.i1 <- 0.5 + 0.3 * sin(angle1)\n",
    "    y.i1 <- 0.5 + 0.35 * cos(angle1)\n",
    "    x.i1 <- x.i1[y.i1 <= max(y.l1)]\n",
    "    y.i1 <- y.i1[y.i1 <= max(y.l1)]\n",
    "    y.i1[1] <- max(y.l1)\n",
    "\n",
    "    x.i2 <- 0.5 + 0.3 * sin(angle2)\n",
    "    y.i2 <- 0.5 + 0.35 * cos(angle2)\n",
    "\n",
    "    x.i <- c(x.i1, x.i2)\n",
    "    y.i <- c(y.i1, y.i2)\n",
    "\n",
    "    x <- c(x.l, 0, 0.2, 0.2, rev(x.i), 0.8, 0.8, 1, 1)\n",
    "    y <- c(y.l, 1, 1, 0.5, rev(y.i), 0.5, 1, 1, 0.85)\n",
    "\n",
    "    x <- .rescale(x)\n",
    "    x <- x.pos + wt * x\n",
    "    y <- y.pos + ht * y\n",
    "\n",
    "    if (is.null(id)) {\n",
    "        id <- rep(1, length(x))\n",
    "    } else {\n",
    "        id <- rep(id, length(x))\n",
    "    }\n",
    "\n",
    "    fill <- rep(fill, 1) # blue\n",
    "\n",
    "    list(x = x, y = y, id = id, fill = fill)\n",
    "}\n",
    "\n",
    "#' Draw C letter\n",
    "#'\n",
    "#' @param x.pos numeric. X position for the letter.\n",
    "#' @param y.pos numeric. Y position for the letter.\n",
    "#' @param ht numeric. Height of the letter.\n",
    "#' @param wt numeric. Width of the letter.\n",
    "#' @param fill character. Color of the letter (polygon shape fill)\n",
    "#' @param id integer. Shape id.\n",
    "#'\n",
    "#' @return list. List with 'x', 'y' coordinates, 'id' and 'fill' of the polygon\n",
    "#' shape (letter).\n",
    "#'\n",
    "#' @keywords internal\n",
    "#' @noRd\n",
    "letterC <- function(x.pos, y.pos, ht, wt, fill = \"#2297E6\", id = NULL) {\n",
    "    angle1 <- seq(0.3 + pi / 2, pi, length = 100)\n",
    "    angle2 <- seq(pi, 1.5 * pi, length = 100)\n",
    "    x.l1 <- 0.5 + 0.5 * sin(angle1)\n",
    "    y.l1 <- 0.5 + 0.5 * cos(angle1)\n",
    "    x.l2 <- 0.5 + 0.5 * sin(angle2)\n",
    "    y.l2 <- 0.5 + 0.5 * cos(angle2)\n",
    "\n",
    "    x.l <- c(x.l1, x.l2)\n",
    "    y.l <- c(y.l1, y.l2)\n",
    "\n",
    "    x <- c(x.l, rev(x.l))\n",
    "    y <- c(y.l, 1 - rev(y.l))\n",
    "\n",
    "    x.i1 <- 0.5 + 0.35 * sin(angle1)\n",
    "    y.i1 <- 0.5 + 0.35 * cos(angle1)\n",
    "    x.i1 <- x.i1[y.i1 <= max(y.l1)]\n",
    "    y.i1 <- y.i1[y.i1 <= max(y.l1)]\n",
    "    y.i1[1] <- max(y.l1)\n",
    "\n",
    "    x.i2 <- 0.5 + 0.35 * sin(angle2)\n",
    "    y.i2 <- 0.5 + 0.35 * cos(angle2)\n",
    "\n",
    "    x.i <- c(x.i1, x.i2)\n",
    "    y.i <- c(y.i1, y.i2)\n",
    "\n",
    "    x1 <- c(x.i, rev(x.i))\n",
    "    y1 <- c(y.i, 1 - rev(y.i))\n",
    "\n",
    "    x <- c(x, rev(x1))\n",
    "    y <- c(y, rev(y1))\n",
    "\n",
    "    x <- .rescale(x)\n",
    "    x <- x.pos + wt * x\n",
    "    y <- y.pos + ht * y\n",
    "\n",
    "    if (is.null(id)) {\n",
    "        id <- rep(1, length(x))\n",
    "    } else {\n",
    "        id <- rep(id, length(x))\n",
    "    }\n",
    "\n",
    "    fill <- rep(fill, 1) # blue\n",
    "\n",
    "    list(x = x, y = y, id = id, fill = fill)\n",
    "}\n",
    "\n",
    "#' Draw G letter\n",
    "#'\n",
    "#' @param x.pos numeric. X position for the letter.\n",
    "#' @param y.pos numeric. Y position for the letter.\n",
    "#' @param ht numeric. Height of the letter.\n",
    "#' @param wt numeric. Width of the letter.\n",
    "#' @param fill character. Color of the letter (polygon shape fill)\n",
    "#' @param id integer. Shape id.\n",
    "#'\n",
    "#' @return list. List with 'x', 'y' coordinates, 'id' and 'fill' of the polygon\n",
    "#' shape (letter).\n",
    "#'\n",
    "#' @keywords internal\n",
    "#' @noRd\n",
    "letterG <- function(x.pos, y.pos, ht, wt, fill = \"#F5C710\", id = NULL) {\n",
    "    angle1 <- seq(0.3 + pi / 2, pi, length = 100)\n",
    "    angle2 <- seq(pi, 1.5 * pi, length = 100)\n",
    "    x.l1 <- 0.5 + 0.5 * sin(angle1)\n",
    "    y.l1 <- 0.5 + 0.5 * cos(angle1)\n",
    "    x.l2 <- 0.5 + 0.5 * sin(angle2)\n",
    "    y.l2 <- 0.5 + 0.5 * cos(angle2)\n",
    "\n",
    "    x.l <- c(x.l1, x.l2)\n",
    "    y.l <- c(y.l1, y.l2)\n",
    "\n",
    "    x <- c(x.l, rev(x.l))\n",
    "    y <- c(y.l, 1 - rev(y.l))\n",
    "\n",
    "    x.i1 <- 0.5 + 0.35 * sin(angle1)\n",
    "    y.i1 <- 0.5 + 0.35 * cos(angle1)\n",
    "    x.i1 <- x.i1[y.i1 <= max(y.l1)]\n",
    "    y.i1 <- y.i1[y.i1 <= max(y.l1)]\n",
    "    y.i1[1] <- max(y.l1)\n",
    "\n",
    "    x.i2 <- 0.5 + 0.35 * sin(angle2)\n",
    "    y.i2 <- 0.5 + 0.35 * cos(angle2)\n",
    "\n",
    "    x.i <- c(x.i1, x.i2)\n",
    "    y.i <- c(y.i1, y.i2)\n",
    "\n",
    "    x1 <- c(x.i, rev(x.i))\n",
    "    y1 <- c(y.i, 1 - rev(y.i))\n",
    "\n",
    "    x <- c(x, rev(x1))\n",
    "    y <- c(y, rev(y1))\n",
    "\n",
    "    h1 <- max(y.l1)\n",
    "    r1 <- max(x.l1)\n",
    "\n",
    "    h1 <- 0.4\n",
    "\n",
    "    x.add <- c(r1, 0.5, 0.5, r1 - 0.2, r1 - 0.2, r1, r1)\n",
    "    y.add <- c(h1, h1, h1 - 0.1, h1 - 0.1, 0, 0, h1)\n",
    "\n",
    "    if (is.null(id)) {\n",
    "        id <- c(rep(1, length(x)), rep(2, length(x.add)))\n",
    "    } else {\n",
    "        id <- c(rep(id, length(x)), rep(id + 1, length(x.add)))\n",
    "    }\n",
    "\n",
    "    x <- c(rev(x), x.add)\n",
    "    y <- c(rev(y), y.add)\n",
    "\n",
    "    x <- .rescale(x)\n",
    "    x <- x.pos + wt * x\n",
    "    y <- y.pos + ht * y\n",
    "\n",
    "\n",
    "    fill <- rep(fill, 2) # orange/yellow\n",
    "\n",
    "    list(x = x, y = y, id = id, fill = fill)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    ".isValidColor <- function(x) {\n",
    "    vapply(x, function(X) {\n",
    "        tryCatch(is.matrix(col2rgb(X)), error = function(e) FALSE)\n",
    "    }, FUN.VALUE = logical(1), USE.NAMES = FALSE)\n",
    "}\n",
    "      \n",
    ".rescale <- function(x, to = c(0.025, 0.975)) {\n",
    "    from <- range(x, na.rm = TRUE, finite = TRUE)\n",
    "    (x - from[1]) / diff(from) * diff(to) + to[1]\n",
    "}\n",
    "           \n",
    "addLetter <- function(letters, which, x.pos, y.pos, ht, wt, fill) {\n",
    "    if (which == \"A\") {\n",
    "        letter <- letterA(x.pos, y.pos, ht, wt, fill = fill[\"A\"])\n",
    "    } else if (which == \"C\") {\n",
    "        letter <- letterC(x.pos, y.pos, ht, wt, fill = fill[\"C\"])\n",
    "    } else if (which == \"G\") {\n",
    "        letter <- letterG(x.pos, y.pos, ht, wt, fill = fill[\"G\"])\n",
    "    } else if (which == \"T\") {\n",
    "        letter <- letterT(x.pos, y.pos, ht, wt, fill = fill[\"T\"])\n",
    "    } else if (which == \"U\") {\n",
    "        letter <- letterU(x.pos, y.pos, ht, wt, fill = fill[\"U\"])\n",
    "    } else {\n",
    "        stop(sprintf(\"\\\"which\\\" must be one of %s\", paste(names(fill), collapse = \", \")))\n",
    "    }\n",
    "\n",
    "    letters$x <- c(letters$x, letter$x)\n",
    "    letters$y <- c(letters$y, letter$y)\n",
    "\n",
    "    lastID <- ifelse(is.null(letters$id), 0, max(letters$id))\n",
    "    letters$id <- c(letters$id, lastID + letter$id)\n",
    "    letters$fill <- c(letters$fill, as.character(letter$fill)) # remove name\n",
    "    letters\n",
    "}\n",
    "           \n",
    "seqLogo2 <- function(pwm, ic.scale = TRUE, xaxis = TRUE, yaxis = TRUE, xfontsize = 15,\n",
    "                    yfontsize = 15, fill = c(A = \"#61D04F\", C = \"#2297E6\", G = \"#F5C710\", T = \"#DF536B\")) {\n",
    "    if (is.data.frame(pwm) || is.matrix(pwm)) {\n",
    "        pwm <- makePWM(pwm)\n",
    "    }\n",
    "\n",
    "    if (!is(pwm, \"pwm\")) {\n",
    "        stop(\"pwm must be of class pwm, matrix or data.frame\")\n",
    "    }\n",
    "\n",
    "    # adjust default color setting if alphabet set to RNA\n",
    "    if (pwm@alphabet == \"RNA\") {\n",
    "        if (\"T\" %in% names(fill)) {\n",
    "            names(fill)[which(names(fill) == \"T\")] <- \"U\"\n",
    "        }\n",
    "    }\n",
    "\n",
    "    if (!is.character(fill) || length(fill) != nrow(pwm@pwm) || !all(rownames(pwm@pwm) %in%\n",
    "        names(fill)) || !all(.isValidColor(fill))) {\n",
    "        stop(\n",
    "            \"\\\"fill\\\" must be named character vector of length \",\n",
    "            sprintf(\n",
    "                \"%d, with names %s (matching the rownames of pwm)\",\n",
    "                nrow(pwm@pwm), paste(rownames(pwm@pwm), collapse = \", \")\n",
    "            ), \"and valid colors as values\"\n",
    "        )\n",
    "    }\n",
    "\n",
    "    chars <- rownames(pwm@pwm)\n",
    "    letters <- list(x = NULL, y = NULL, id = NULL, fill = NULL)\n",
    "    npos <- ncol(pwm@pwm)\n",
    "\n",
    "\n",
    "    if (ic.scale) {\n",
    "        ylim <- 2\n",
    "        ylab <- \"Information content\"\n",
    "        facs <- pwm@ic\n",
    "    } else {\n",
    "        ylim <- 1\n",
    "        ylab <- \"Probability\"\n",
    "        facs <- rep(1, npos)\n",
    "    }\n",
    "\n",
    "    wt <- 1\n",
    "    x.pos <- 0\n",
    "    for (j in seq_len(npos)) {\n",
    "        column <- pwm@pwm[, j]\n",
    "        hts <- 0.95 * column * facs[j]\n",
    "        letterOrder <- order(hts)\n",
    "\n",
    "        y.pos <- 0\n",
    "        for (i in seq_len(4)) {\n",
    "            letter <- chars[letterOrder[i]]\n",
    "            ht <- hts[letterOrder[i]]\n",
    "            if (ht > 0) {\n",
    "                  letters <- addLetter(letters, letter, x.pos, y.pos, ht, wt, fill)\n",
    "              }\n",
    "            y.pos <- y.pos + ht + 0.01\n",
    "        }\n",
    "        x.pos <- x.pos + wt\n",
    "    }\n",
    "\n",
    "    grid.newpage()\n",
    "    bottomMargin <- ifelse(xaxis, 2 + xfontsize / 3.5, 2)\n",
    "    leftMargin <- ifelse(yaxis, 2 + yfontsize / 3.5, 2)\n",
    "    pushViewport(plotViewport(c(bottomMargin, leftMargin, 2, 2)))\n",
    "    pushViewport(dataViewport(0:ncol(pwm@pwm), 0:ylim, name = \"vp1\"))\n",
    "    grid.polygon(\n",
    "        x = unit(letters$x, \"native\"), y = unit(letters$y, \"native\"), id = letters$id,\n",
    "        gp = gpar(fill = letters$fill, col = \"transparent\")\n",
    "    )\n",
    "    if (xaxis) {\n",
    "        grid.xaxis(\n",
    "            at = seq(0.5, ncol(pwm@pwm) - 0.5), label = seq_len(ncol(pwm@pwm)),\n",
    "            gp = gpar(fontsize = xfontsize)\n",
    "        )\n",
    "        grid.text(\"Position\", y = unit(-3, \"lines\"), gp = gpar(fontsize = xfontsize))\n",
    "    }\n",
    "    if (yaxis) {\n",
    "        grid.yaxis(gp = gpar(fontsize = yfontsize))\n",
    "        grid.text(ylab, x = unit(-3, \"lines\"), rot = 90, gp = gpar(fontsize = yfontsize))\n",
    "    }\n",
    "    popViewport()\n",
    "    popViewport()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
