{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "904183a2",
   "metadata": {},
   "source": [
    "authors: Eric Kernfeld, Macrina Lobo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8492847f",
   "metadata": {},
   "source": [
    "- install pacakges; more install done from terminal elsewhere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a939e45f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//RtmpesRfwI/downloaded_packages\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"here\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f57e86d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocManager\n",
      "\n",
      "Bioconductor version '3.10' is out-of-date; the current release version '3.13'\n",
      "  is available with R version '4.1'; see https://bioconductor.org/install\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.10 (BiocManager 1.30.16), R 3.6.3 (2020-02-29)\n",
      "\n",
      "Installing github package(s) 'tanaylab/metacell'\n",
      "\n",
      "Downloading GitHub repo tanaylab/metacell@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 method overwritten by 'cli':\n",
      "  method     from         \n",
      "  print.boxx spatstat.geom\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking for file ‘/private/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T/RtmpesRfwI/remotes594d53a63783/tanaylab-metacell-e6b15d2/DESCRIPTION’\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mpreparing ‘metacell’:\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking DESCRIPTION meta-information\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for LF line-endings in source and make files and shell scripts\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for empty or unneeded directories\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mbuilding ‘metacell_0.3.6.tar.gz’\u001b[39m\u001b[36m\u001b[39m\n",
      "   \n",
      "\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'gam', 'quantreg', 'readr'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!require(\"BiocManager\")) install.packages('BiocManager') \n",
    "BiocManager::install(\"tanaylab/metacell\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "758feb44",
   "metadata": {},
   "source": [
    "- load packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "859000b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"Seurat\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f157feb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "here() starts at /Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"Matrix\")\n",
    "  library(\"magrittr\")\n",
    "\n",
    "\n",
    "  library(\"here\")\n",
    "\n",
    "  library(\"metacell\")\n",
    "  library(\"parallel\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "409b94ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(SCENIC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0ed55dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "help(\"metacell\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "763f077d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "here() starts at /Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts.\n",
      "- This directory and its parents do not satisfy any of the following criteria:\n",
      "  - contains a file \".here\"\n",
      "  - contains a file matching \"[.]Rproj$\" with contents matching \"^Version: \" in the first line\n",
      "  - contains a file \"DESCRIPTION\" with contents matching \"^Package: \"\n",
      "  - contains a file \"remake.yml\"\n",
      "  - contains a file \".projectile\"\n",
      "  - contains a directory \".git\"\n",
      "  - contains a file \".git\" with contents matching \"^gitdir: \"\n",
      "  - contains a directory \".svn\"\n",
      "! Call `here::i_am()` in your scripts and reports, or use `set_here()` to create a `.here` file\n",
      "- Initial working directory: /Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts\n",
      "- Current working directory: /Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "'/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts'"
      ],
      "text/latex": [
       "'/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass\\_Macrina\\_Rene/rerun\\_scenic/scripts'"
      ],
      "text/markdown": [
       "'/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts'"
      ],
      "text/plain": [
       "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts'"
      ],
      "text/latex": [
       "'/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass\\_Macrina\\_Rene/rerun\\_scenic/scripts'"
      ],
      "text/markdown": [
       "'/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts'"
      ],
      "text/plain": [
       "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dr_here()\n",
    "  PROJ_DIR = here()\n",
    "  getwd()\n",
    "  here()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "948eb532",
   "metadata": {},
   "source": [
    "- more installs  SCENIC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0fc892da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] ‘3.10’"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.10 (BiocManager 1.30.12), R 3.6.3 (2020-02-29)\n",
      "\n",
      "Installing package(s) 'AUCell', 'RcisTarget'\n",
      "\n",
      "also installing the dependencies ‘GSEABase’, ‘feather’\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//Rtmp41rDZu/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'ade4', 'ape', 'beeswarm', 'BiocManager', 'boot', 'brio',\n",
      "  'callr', 'car', 'circlize', 'class', 'cluster', 'compositions', 'corrplot',\n",
      "  'curl', 'dendextend', 'DEoptimR', 'devtools', 'e1071', 'fitdistrplus',\n",
      "  'formatR', 'gam', 'gert', 'GGally', 'ggsignif', 'gh', 'gtools', 'haven',\n",
      "  'highr', 'htmlTable', 'IRkernel', 'KernSmooth', 'knitr', 'ks', 'lattice',\n",
      "  'leiden', 'lme4', 'MASS', 'Matrix', 'mgcv', 'mime', 'multicool', 'mvtnorm',\n",
      "  'nnet', 'openxlsx', 'parallelly', 'plot3D', 'plotly', 'processx', 'proxy',\n",
      "  'quantreg', 'RcppArmadillo', 'remotes', 'rhandsontable', 'rio', 'rmarkdown',\n",
      "  'robustbase', 'rversions', 'segmented', 'seqinr', 'seriation', 'spatial',\n",
      "  'spatstat', 'spatstat.core', 'spatstat.geom', 'spatstat.linnet',\n",
      "  'spatstat.utils', 'statmod', 'survival', 'testthat', 'tidyverse', 'tinytex',\n",
      "  'xfun', 'zip'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.10 (BiocManager 1.30.12), R 3.6.3 (2020-02-29)\n",
      "\n",
      "Installing package(s) 'GENIE3'\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//Rtmp41rDZu/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'ade4', 'ape', 'beeswarm', 'BiocManager', 'boot', 'brio',\n",
      "  'callr', 'car', 'circlize', 'class', 'cluster', 'compositions', 'corrplot',\n",
      "  'curl', 'dendextend', 'DEoptimR', 'devtools', 'e1071', 'fitdistrplus',\n",
      "  'formatR', 'gam', 'gert', 'GGally', 'ggsignif', 'gh', 'gtools', 'haven',\n",
      "  'highr', 'htmlTable', 'IRkernel', 'KernSmooth', 'knitr', 'ks', 'lattice',\n",
      "  'leiden', 'lme4', 'MASS', 'Matrix', 'mgcv', 'mime', 'multicool', 'mvtnorm',\n",
      "  'nnet', 'openxlsx', 'parallelly', 'plot3D', 'plotly', 'processx', 'proxy',\n",
      "  'quantreg', 'RcppArmadillo', 'remotes', 'rhandsontable', 'rio', 'rmarkdown',\n",
      "  'robustbase', 'rversions', 'segmented', 'seqinr', 'seriation', 'spatial',\n",
      "  'spatstat', 'spatstat.core', 'spatstat.geom', 'spatstat.linnet',\n",
      "  'spatstat.utils', 'statmod', 'survival', 'testthat', 'tidyverse', 'tinytex',\n",
      "  'xfun', 'zip'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.10 (BiocManager 1.30.12), R 3.6.3 (2020-02-29)\n",
      "\n",
      "Installing package(s) 'zoo', 'mixtools', 'rbokeh'\n",
      "\n",
      "Warning message:\n",
      "“package ‘rbokeh’ is not available (for R version 3.6.3)”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//Rtmp41rDZu/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'ade4', 'ape', 'beeswarm', 'BiocManager', 'boot', 'brio',\n",
      "  'callr', 'car', 'circlize', 'class', 'cluster', 'compositions', 'corrplot',\n",
      "  'curl', 'dendextend', 'DEoptimR', 'devtools', 'e1071', 'fitdistrplus',\n",
      "  'formatR', 'gam', 'gert', 'GGally', 'ggsignif', 'gh', 'gtools', 'haven',\n",
      "  'highr', 'htmlTable', 'IRkernel', 'KernSmooth', 'knitr', 'ks', 'lattice',\n",
      "  'leiden', 'lme4', 'MASS', 'Matrix', 'mgcv', 'mime', 'multicool', 'mvtnorm',\n",
      "  'nnet', 'openxlsx', 'parallelly', 'plot3D', 'plotly', 'processx', 'proxy',\n",
      "  'quantreg', 'RcppArmadillo', 'remotes', 'rhandsontable', 'rio', 'rmarkdown',\n",
      "  'robustbase', 'rversions', 'segmented', 'seqinr', 'seriation', 'spatial',\n",
      "  'spatstat', 'spatstat.core', 'spatstat.geom', 'spatstat.linnet',\n",
      "  'spatstat.utils', 'statmod', 'survival', 'testthat', 'tidyverse', 'tinytex',\n",
      "  'xfun', 'zip'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.10 (BiocManager 1.30.12), R 3.6.3 (2020-02-29)\n",
      "\n",
      "Installing package(s) 'DT', 'NMF', 'ComplexHeatmap', 'R2HTML', 'Rtsne'\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//Rtmp41rDZu/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'ade4', 'ape', 'beeswarm', 'BiocManager', 'boot', 'brio',\n",
      "  'callr', 'car', 'circlize', 'class', 'cluster', 'compositions', 'corrplot',\n",
      "  'curl', 'dendextend', 'DEoptimR', 'devtools', 'e1071', 'fitdistrplus',\n",
      "  'formatR', 'gam', 'gert', 'GGally', 'ggsignif', 'gh', 'gtools', 'haven',\n",
      "  'highr', 'htmlTable', 'IRkernel', 'KernSmooth', 'knitr', 'ks', 'lattice',\n",
      "  'leiden', 'lme4', 'MASS', 'Matrix', 'mgcv', 'mime', 'multicool', 'mvtnorm',\n",
      "  'nnet', 'openxlsx', 'parallelly', 'plot3D', 'plotly', 'processx', 'proxy',\n",
      "  'quantreg', 'RcppArmadillo', 'remotes', 'rhandsontable', 'rio', 'rmarkdown',\n",
      "  'robustbase', 'rversions', 'segmented', 'seqinr', 'seriation', 'spatial',\n",
      "  'spatstat', 'spatstat.core', 'spatstat.geom', 'spatstat.linnet',\n",
      "  'spatstat.utils', 'statmod', 'survival', 'testthat', 'tidyverse', 'tinytex',\n",
      "  'xfun', 'zip'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.10 (BiocManager 1.30.12), R 3.6.3 (2020-02-29)\n",
      "\n",
      "Installing package(s) 'doMC', 'doRNG'\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//Rtmp41rDZu/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'ade4', 'ape', 'beeswarm', 'BiocManager', 'boot', 'brio',\n",
      "  'callr', 'car', 'circlize', 'class', 'cluster', 'compositions', 'corrplot',\n",
      "  'curl', 'dendextend', 'DEoptimR', 'devtools', 'e1071', 'fitdistrplus',\n",
      "  'formatR', 'gam', 'gert', 'GGally', 'ggsignif', 'gh', 'gtools', 'haven',\n",
      "  'highr', 'htmlTable', 'IRkernel', 'KernSmooth', 'knitr', 'ks', 'lattice',\n",
      "  'leiden', 'lme4', 'MASS', 'Matrix', 'mgcv', 'mime', 'multicool', 'mvtnorm',\n",
      "  'nnet', 'openxlsx', 'parallelly', 'plot3D', 'plotly', 'processx', 'proxy',\n",
      "  'quantreg', 'RcppArmadillo', 'remotes', 'rhandsontable', 'rio', 'rmarkdown',\n",
      "  'robustbase', 'rversions', 'segmented', 'seqinr', 'seriation', 'spatial',\n",
      "  'spatstat', 'spatstat.core', 'spatstat.geom', 'spatstat.linnet',\n",
      "  'spatstat.utils', 'statmod', 'survival', 'testthat', 'tidyverse', 'tinytex',\n",
      "  'xfun', 'zip'\n",
      "\n",
      "Downloading GitHub repo aertslab/SCopeLoomR@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rlist (NA -> 0.4.6.1) [CRAN]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing 1 packages: rlist\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//Rtmp41rDZu/downloaded_packages\n",
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking for file ‘/private/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T/Rtmp41rDZu/remotes26082143c34f/aertslab-SCopeLoomR-f065eae/DESCRIPTION’\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mpreparing ‘SCopeLoomR’:\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking DESCRIPTION meta-information\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for LF line-endings in source and make files and shell scripts\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for empty or unneeded directories\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mlooking to see if a ‘data/datalist’ file should be added\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mbuilding ‘SCopeLoomR_0.11.0.tar.gz’\u001b[39m\u001b[36m\u001b[39m\n",
      "   \n",
      "\r"
     ]
    }
   ],
   "source": [
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE)) install.packages(\"BiocManager\")\n",
    "BiocManager::version()\n",
    "# If your bioconductor version is previous to 4.0, see the section bellow\n",
    "\n",
    "## Required\n",
    "BiocManager::install(c(\"AUCell\", \"RcisTarget\"))\n",
    "BiocManager::install(c(\"GENIE3\")) # Optional. Can be replaced by GRNBoost\n",
    "\n",
    "## Optional (but highly recommended):\n",
    "# To score the network on cells (i.e. run AUCell):\n",
    "BiocManager::install(c(\"zoo\", \"mixtools\", \"rbokeh\"))\n",
    "# For various visualizations and perform t-SNEs:\n",
    "BiocManager::install(c(\"DT\", \"NMF\", \"ComplexHeatmap\", \"R2HTML\", \"Rtsne\"))\n",
    "# To support paralell execution (not available in Windows):\n",
    "BiocManager::install(c(\"doMC\", \"doRNG\"))\n",
    "# To export/visualize in http://scope.aertslab.org\n",
    "if (!requireNamespace(\"devtools\", quietly = TRUE)) install.packages(\"devtools\")\n",
    "devtools::install_github(\"aertslab/SCopeLoomR\", build_vignettes = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1582518a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading GitHub repo aertslab/SCENIC@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking for file ‘/private/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T/Rtmp41rDZu/remotes260815e48cd9/aertslab-SCENIC-0585e87/DESCRIPTION’\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mpreparing ‘SCENIC’:\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking DESCRIPTION meta-information\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for LF line-endings in source and make files and shell scripts\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for empty or unneeded directories\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mlooking to see if a ‘data/datalist’ file should be added\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mbuilding ‘SCENIC_1.2.4.tar.gz’\u001b[39m\u001b[36m\u001b[39m\n",
      "   \n",
      "\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in i.p(...):\n",
      "“installation of package ‘/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//Rtmp41rDZu/file26084fdb05c9/SCENIC_1.2.4.tar.gz’ had non-zero exit status”\n"
     ]
    }
   ],
   "source": [
    "if (!requireNamespace(\"devtools\", quietly = TRUE)) install.packages(\"devtools\")\n",
    "devtools::install_github(\"aertslab/SCENIC\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ceca76ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//Rtmp41rDZu/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading GitHub repo aertslab/SCENIC@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking for file ‘/private/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T/Rtmp41rDZu/remotes260864dad117/aertslab-SCENIC-0585e87/DESCRIPTION’\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mpreparing ‘SCENIC’:\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking DESCRIPTION meta-information\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for LF line-endings in source and make files and shell scripts\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for empty or unneeded directories\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mlooking to see if a ‘data/datalist’ file should be added\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mbuilding ‘SCENIC_1.2.4.tar.gz’\u001b[39m\u001b[36m\u001b[39m\n",
      "   \n",
      "\r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in i.p(...):\n",
      "“installation of package ‘/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//Rtmp41rDZu/file260859d570bb/SCENIC_1.2.4.tar.gz’ had non-zero exit status”\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"remotes\")\n",
    "remotes::install_github(\"aertslab/SCENIC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "01475bf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# setup motif analysis files here, mm9\n",
    "aerts_lab_cistargets = \"https://resources.aertslab.org/cistarget/databases/mus_musculus/mm9/refseq_r45/mc9nr/gene_based\"\n",
    "dbFiles <- c(\"mm9-500bp-upstream-7species.mc9nr\", \"mm9-tss-centered-10kb-7species.mc9nr\")\n",
    "\n",
    "suffix =\".feather\"\n",
    "for(featherURL in dbFiles){\n",
    "  filename = paste0(featherURL, suffix)\n",
    "  if(!file.exists(filename)){\n",
    "    download.file(file.path(aerts_lab_cistargets, filename), destfile=filename) \n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fd34269",
   "metadata": {},
   "source": [
    "- required mm10 files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "22f90008",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in download.file(file.path(aerts_lab_cistargets, filename), destfile = filename):\n",
      "“cannot open URL 'https://resources.aertslab.org/cistarget/databases/mus_musculus/mm10/refseq_r80/mc9nr/gene_based/mm9-500bp-upstream-7species.mc9nr.feather': HTTP status was '404 Not Found'”\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in download.file(file.path(aerts_lab_cistargets, filename), destfile = filename): cannot open URL 'https://resources.aertslab.org/cistarget/databases/mus_musculus/mm10/refseq_r80/mc9nr/gene_based/mm9-500bp-upstream-7species.mc9nr.feather'\n",
     "output_type": "error",
     "traceback": [
      "Error in download.file(file.path(aerts_lab_cistargets, filename), destfile = filename): cannot open URL 'https://resources.aertslab.org/cistarget/databases/mus_musculus/mm10/refseq_r80/mc9nr/gene_based/mm9-500bp-upstream-7species.mc9nr.feather'\nTraceback:\n",
      "1. download.file(file.path(aerts_lab_cistargets, filename), destfile = filename)"
     ]
    }
   ],
   "source": [
    "# mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr\n",
    "# mm10__refseq-r80__10kb_up_and_down_tss.mc9nr\n",
    "\n",
    "# setup motif analysis files here, mm9\n",
    "aerts_lab_cistargets = \"https://resources.aertslab.org/cistarget/databases/mus_musculus/mm10/refseq_r80/mc9nr/gene_based\"\n",
    "dbFiles <- c(\"mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr\", \"mm10__refseq-r80__10kb_up_and_down_tss.mc9nr\")\n",
    "\n",
    "suffix =\".feather\"\n",
    "for(featherURL in dbFiles){\n",
    "  filename = paste0(featherURL, suffix)\n",
    "  if(!file.exists(filename)){\n",
    "    download.file(file.path(aerts_lab_cistargets, filename), destfile=filename) \n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "adb7b422",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/hm/9slqvx2n1_q0y184ts0ck7sc0000gp/T//Rtmp5itBLq/downloaded_packages\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"arrow\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe701df8",
   "metadata": {},
   "source": [
    "- functions, misc setup modified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5edcc5bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "org = \"mgi\" # or hgnc, or dmel\n",
    "dbDir = \"/Users/LoboM/Dropbox/sharedUMass_Macrina_Rene/rerun_scenic/scripts/\" # RcisTarget databases location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "7d03feb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Motif databases selected: \n",
      "  mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.feather \n",
      "  mm10__refseq-r80__10kb_up_and_down_tss.mc9nr.feather \n",
      "  mm9-500bp-upstream-7species.mc9nr.feather \n",
      "  mm9-tss-centered-10kb-7species.mc9nr.feather\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# excluded mm9 for the paper\n",
    "scenicOptions <- initializeScenic(org=org, dbs= c(\"mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.feather\", \"mm10__refseq-r80__10kb_up_and_down_tss.mc9nr.feather\",\"mm9-500bp-upstream-7species.mc9nr.feather\", \"mm9-tss-centered-10kb-7species.mc9nr.feather\"),dbDir=dbDir, datasetTitle=\"scenic_pharynx\", nCores=6) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ea54daf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "PERMITTED_DGE_TYPES = list(\n",
    "  dge_type = c(\"metacell\", \"full\", \"velocyto_loom\", \n",
    "               \"velocyto_spliced\",   \"velocyto_spliced_metacell\", \n",
    "               \"velocyto_unspliced\", \"velocyto_unspliced_metacell\"),\n",
    "  corresponding_metadata_type = c(\"metacell\", \"full\", \"full\", \n",
    "                                  \"full\", \"metacell\", \n",
    "                                  \"full\", \"metacell\"),\n",
    "  metacell_precursors     = c(\"full\",     \"velocyto_spliced\",          \"velocyto_unspliced\"), \n",
    "  corresponding_metacells = c(\"metacell\", \"velocyto_spliced_metacell\", \"velocyto_unspliced_metacell\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "01e36d9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_metadata_type = function(dge_type) {\n",
    "  # assertthat::assert_that(dge_type %in% PERMITTED_DGE_TYPES$dge_type)\n",
    "  setNames( PERMITTED_DGE_TYPES$corresponding_metadata_type, PERMITTED_DGE_TYPES$dge_type)[[dge_type]]\n",
    "}\n",
    "get_metacell_type = function(dge_type){\n",
    "  # assertthat::assert_that(dge_type %in% PERMITTED_DGE_TYPES$metacell_precursors)\n",
    "  setNames( PERMITTED_DGE_TYPES$corresponding_metacells, PERMITTED_DGE_TYPES$metacell_precursors)[[dge_type]]\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "54b310b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Load cell metadata\n",
    "#' \n",
    "#' @param type full data or metacell?\n",
    "#' \n",
    "#' Returns a dataframe\n",
    "load_cell_metadata = function( dataset_name = NULL, type = \"full\", stop_for_missing_file = T ){\n",
    "  \n",
    "    # assertthat::assert_that(type %in% PERMITTED_DGE_TYPES$dge_type)\n",
    "  \n",
    "  file_path = here(\"data\", dataset_name, paste0(get_metadata_type(type), \"_metadata.csv\"))\n",
    "  print(file_path)\n",
    "  if(!file.exists(file_path) ){\n",
    "    if(stop_for_missing_file){\n",
    "      stop(\": Metadata not available.\\n\" %>% paste0(dataset_name, .))\n",
    "    } else {\n",
    "      return()\n",
    "    }\n",
    "  }\n",
    "  metadata_df = read.csv(file_path, stringsAsFactors = F)\n",
    "  rownames(metadata_df) = metadata_df[[\"CELL_ID\"]]\n",
    "  metadata_df\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "73a8e1da",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Apply a function to all dge's from a given source. \n",
    "#' \n",
    "#' @param dataset_name Which data source; e.g. \"pharynx_atlas\"\n",
    "#' @param test If true, the just run on a randomly selected DGE. If false, run on all of them.\n",
    "#' @param type \"full\" for cells, \"metacell\" for metacells\n",
    "#' @param omit Skip these.\n",
    "#' @param group_by If you want to group by a different variable, for example \"BATCH\". This \n",
    "#' may become more flexible in the future, but for now you can only group by a variable\n",
    "#' in which DGE is nested (a strictly \"coarser\" variable). \n",
    "#' \n",
    "iterate_within_source = function(dataset_name, FUN, \n",
    "                                 test = F,\n",
    "                                 omit = NULL, \n",
    "                                 group_by = \"DGE\", \n",
    "                                 ...){\n",
    "  X = load_cell_metadata(dataset_name) \n",
    "  X = X[!(X[[group_by]] %in% omit), ]\n",
    "  available_dges = X %>% extract(c(group_by, \"DGE\") ) %>%\n",
    "    table %>% as.data.frame %>% subset(Freq > 0) %>% \n",
    "    (plyr::dlply)(.variables = group_by, .fun = extract2, \"DGE\") %>% \n",
    "    lapply(as.character)\n",
    "  \n",
    "  if(length(available_dges)==0){\n",
    "    warning(paste0(\"No expression data available: \", dataset_name, \" type = \", type))\n",
    "    return()\n",
    "  }\n",
    "  if(test){\n",
    "    available_dges = sample(available_dges, 1)\n",
    "  }\n",
    "  ma = list(...)\n",
    "  print(available_dges)\n",
    "  mapply(FUN, dataset_name, available_dges, MoreArgs = ma, SIMPLIFY = F) %>% setNames(available_dges)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "9cdc4856",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Load expression data from a given dge matrix.\n",
    "#' \n",
    "#' @param dataset_name A subfolder of \"data/\".\n",
    "#' @param dge_name A dge in that subfolder. Can come from the DGE field of the cell metadata.\n",
    "#' @param type Try \"full\", \"metacell\", or anything in PERMITTED_DGE_TYPES$dge_type .\n",
    "#' @param fix_cellnames Cell ID's get passed through this function. \n",
    "#' @param mode If simple, return what you see in the file. \n",
    "#' If helpful, look for cells missing from metadata. If any found, throw a warning. They are assumed to be empties or low quality and they are not returned.\n",
    "#' If strict, look for cells missing from metadata. If any found, throw an error.\n",
    "#' Returns a Matrix\n",
    "load_expr_by_dge = function( dataset_name, dge_name, type = \"full\", mode, fix_cellnames = NULL, ... ){\n",
    "  assertthat::assert_that(mode %in% c(\"simple\", \"helpful\", \"strict\"))\n",
    "  assertthat::assert_that(type %in% PERMITTED_DGE_TYPES$dge_type)\n",
    "  \n",
    "  # Find it\n",
    "  dge_location = here(\"data\", dataset_name, type %>% gsub('(un)*spliced$', \"loom\", .), dge_name) \n",
    "  print(dge_location)\n",
    "  if( type %in% c(\"velocyto_loom\", \"velocyto_spliced\", \"velocyto_unspliced\")){\n",
    "    dge_location %<>% paste0(\".loom\")\n",
    "  }\n",
    "  if(is.null(dge_location)){\n",
    "    stop(\"Error locating expression matrix! Please report this error to Eric.\\n\")\n",
    "  }\n",
    "  \n",
    "  # read anything -- loom, .tsv, .csv, 10Xv2, 10Xv3\n",
    "  X = read_one_dge_flexibly(dge_location, ...)\n",
    "  \n",
    "  # deal with velocity-associated bullshit\n",
    "  if( type == \"velocyto_spliced\" ){\n",
    "    X = X[[\"spliced\"]]\n",
    "  }\n",
    "  if( type == \"velocyto_unspliced\" ){\n",
    "    X = X[[\"unspliced\"]]\n",
    "  }\n",
    "  if( !is.null( fix_cellnames )){\n",
    "    colnames(X) %<>% fix_cellnames\n",
    "  }\n",
    "  \n",
    "  # Return cells depending on mode\n",
    "  if(mode==\"simple\"){\n",
    "    return(X)\n",
    "  }\n",
    "  metadata = load_cell_metadata(dataset_name = dataset_name, type = type)\n",
    "  good_cells = intersect(colnames(X), metadata$CELL_ID)\n",
    "  if( !identical( good_cells, colnames(X)) ){\n",
    "    if(mode==\"helpful\"){\n",
    "      y = length(colnames(X))\n",
    "      x = y - length(good_cells)\n",
    "      warning(\n",
    "        paste( x, \"of\", y, \n",
    "               \"cells missing from metadata! Will omit those from expr matrix.\\n\",\n",
    "               dataset_name, \"   \", dge_name)\n",
    "              )\n",
    "    } \n",
    "    if(mode==\"strict\"){\n",
    "      check_containment(colnames(X), metadata$CELL_ID, error_message = \"Cells missing from metadata!\")\n",
    "      stop(\"Cells missing from metadata!\\n\")\n",
    "    }\n",
    "  }\n",
    "  return(X[,good_cells])\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ba59fd7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "atat = function(my_assertion) (assertthat::assert_that(my_assertion))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fe361866",
   "metadata": {},
   "outputs": [],
   "source": [
    "## ------------------------------------------------------------------------\n",
    "#' Merge a list of digital gene expression matrices.\n",
    "#'\n",
    "#' @param dge_list list of matrices with genes as rows and cell barcodes as columns. \n",
    "#' @param allow_barcode_overlap Set this to TRUE if matching barcodes in different inputs refer to the same cells, \n",
    "#' as in a resequencing of a library. Set this to FALSE (default) if matching barcodes would only occur by coincidence, \n",
    "#' as in merging of different replicates for joint analysis.\n",
    "#' @details `dge_merge_list` converts the digital gene expression matrices to dataframes with genes as columns, merges them, then converts the result back, all without disturbing the gene labels. A gene will be included if it appears in any of the datasets. If a gene appears in one dataset but not another, zeroes will be filled in for missing expression levels.\n",
    "#' @export\n",
    "dge_merge_list = function(dge_list, allow_barcode_overlap = F){\n",
    "  # Allow duplicate cells but not duplicate genes\n",
    "  all_genes = Reduce( f = union, x = lapply( dge_list, rownames ) )\n",
    "  all_cells = Reduce( f = c,     x = lapply( dge_list, colnames ) )\n",
    "  input_dimensions = Reduce(rbind, lapply(dge_list, dim)) %>% matrix(ncol = 2)\n",
    "\n",
    "  if( allow_barcode_overlap ){\n",
    "    all_cells = unique( all_cells )\n",
    "  } else {\n",
    "    if(anyDuplicated(all_cells)){ \n",
    "      warning( \"These samples contain overlapping barcodes! \n",
    "                \\nAttempting to append names(dge_list) to barcodes...\" )\n",
    "      if(is.null(names(dge_list))){ \n",
    "        stop(\"...No names available for dge_list!\")\n",
    "      }\n",
    "      all_cells = lapply( dge_list, colnames ) %>% \n",
    "        mapply( paste0, .,  \"|\", names(dge_list) ) %>%\n",
    "        Reduce( f = c,     x = . )\n",
    "    }\n",
    "    atat(!anyDuplicated(all_cells)) \n",
    "  }\n",
    "  \n",
    "  new_dge = Matrix::Matrix( 0, \n",
    "                            nrow = length( all_genes ), \n",
    "                            ncol = length( all_cells ), \n",
    "                            sparse = T ) \n",
    "  rownames( new_dge ) = all_genes\n",
    "  colnames( new_dge ) = all_cells\n",
    "  for( ii in seq_along( dge_list ) ){\n",
    "    rr = rownames(dge_list[[ii]])\n",
    "    cc = colnames(dge_list[[ii]])\n",
    "    new_dge[rr, cc] = new_dge[rr, cc] + dge_list[[ii]]\n",
    "    print( paste( \"Finished merging dge matrix\", names(dge_list)[[ii]] ) )\n",
    "    dge_list[[ii]] = matrix(); gc()\n",
    "  }\n",
    "  #Results should have at least as many genes (barcodes) as the input with the most genes (barcodes) \n",
    "  assertthat::assert_that(dim(new_dge)[1] >= max(input_dimensions[,1]))\n",
    "  assertthat::assert_that(dim(new_dge)[2] >= max(input_dimensions[,2]))\n",
    "  #If no barcode overlap, should retain all input barcodes.\n",
    "  if( !allow_barcode_overlap ){\n",
    "    assertthat::assert_that(dim(new_dge)[2] == sum(input_dimensions[,2]))\n",
    "  }\n",
    "  assertthat::assert_that(!is.null(colnames(new_dge)))\n",
    "  assertthat::assert_that(!is.null(rownames(new_dge)))\n",
    "  return( new_dge )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "56bbef12",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Seurat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27653765",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2c3a42fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "Read10X = function (data.dir = NULL, version = NULL) \n",
    "{\n",
    "  version = ifelse(file.exists(file.path(data.dir, \"features.tsv.gz\")), 3, 2)\n",
    "  full.data <- list()\n",
    "  for(i in seq_along(data.dir)){\n",
    "    run <- data.dir[i]\n",
    "    if (!dir.exists(run)) {\n",
    "      stop(\"Directory provided does not exist\")\n",
    "    }\n",
    "    if (!grepl(\"\\\\/$\", run)) {\n",
    "      run <- paste(run, \"/\", sep = \"\")\n",
    "    }\n",
    "    if( version == 3 ){\n",
    "      barcode.loc <- paste0(run, \"barcodes.tsv.gz\")\n",
    "      gene.loc <- paste0(run, \"features.tsv.gz\")\n",
    "      matrix.loc <- paste0(run, \"matrix.mtx.gz\")\n",
    "      modifier = function(x) gzcon(file(x, open = \"r\"))\n",
    "    } else if (version == 2){\n",
    "      barcode.loc <- paste0(run, \"barcodes.tsv\")\n",
    "      gene.loc <- paste0(run, \"genes.tsv\")\n",
    "      matrix.loc <- paste0(run, \"matrix.mtx\")\n",
    "      modifier = function(x) x\n",
    "    }\n",
    "    if (!file.exists(barcode.loc)) {\n",
    "      stop(\"Barcode file missing\")\n",
    "    }\n",
    "    if (!file.exists(gene.loc)) {\n",
    "      stop(\"Gene name file missing\")\n",
    "    }\n",
    "    if (!file.exists(matrix.loc)) {\n",
    "      stop(\"Expression matrix file missing\")\n",
    "    }\n",
    "    matrix.loc = modifier(matrix.loc)\n",
    "    barcode.loc = modifier(barcode.loc)\n",
    "    gene.loc = modifier(gene.loc)\n",
    "    data <- Matrix::readMM(file = matrix.loc)\n",
    "    cell.names <- readLines(barcode.loc)\n",
    "    gene.names <- readLines(gene.loc)\n",
    "    try(close(matrix.loc), silent = T)\n",
    "    try(close(barcode.loc), silent = T)\n",
    "    try(close(gene.loc), silent = T)\n",
    "    \n",
    "    print(cell.names)\n",
    "    if (all(grepl(pattern = \"\\\\-1$\", x = cell.names))) {\n",
    "      cell.names <- as.vector(x = as.character(x = sapply(X = cell.names, \n",
    "                                                          FUN = ExtractField_me, \n",
    "                                                          field = 1,\n",
    "                                                          delim = \"-\")))\n",
    "    }\n",
    "    # print(cell.names)\n",
    "    #   print(gene.names)\n",
    "    rownames(x = data) <- make.unique(names = as.character(x = sapply(X = gene.names,\n",
    "                                                                      FUN = ExtractField_me, \n",
    "                                                                      field = 2,\n",
    "                                                                      delim = \"\\\\t\")))\n",
    "    if (is.null(x = names(x = data.dir))) {\n",
    "      if (i < 2) {\n",
    "        colnames(x = data) <- cell.names\n",
    "      }\n",
    "      else {\n",
    "        colnames(x = data) <- paste0(i, \"_\", cell.names)\n",
    "      }\n",
    "    }\n",
    "    else {\n",
    "      colnames(x = data) <- paste0(names(x = data.dir)[i], \n",
    "                                   \"_\", cell.names)\n",
    "    }\n",
    "    full.data <- append(x = full.data, values = data)\n",
    "    full.data <- do.call(cbind, full.data)\n",
    "    return(full.data)\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "acc36d62",
   "metadata": {},
   "outputs": [],
   "source": [
    "ExtractField_me <- function(string, field = 1, delim = \"_\") {\n",
    "  fields <- as.numeric(x = unlist(x = strsplit(x = as.character(x = field), split = \",\")))\n",
    "  if (length(x = fields) == 1) {\n",
    "    return(strsplit(x = string, split = delim)[[1]][field])\n",
    "  }\n",
    "  return(paste(strsplit(x = string, split = delim)[[1]][fields], collapse = delim))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "d4a1d94f",
   "metadata": {},
   "outputs": [],
   "source": [
    "read_one_dge_flexibly = function(to_read, min_umi_count = 100){\n",
    "  extension = tools::file_ext(gsub(\".gz\", \"\", to_read))\n",
    "  if (extension %in% c(\"txt\", \"tsv\")){\n",
    "    X = read.table(to_read, header = T, row.names = 1) %>% as.matrix %>% Matrix(sparse = T)\n",
    "  } else if (extension==\"loom\"){\n",
    "    X = velocyto.R::read.loom.matrices(to_read)\n",
    "    if( min_umi_count!=100 ){ warning(\"min_umi_count is ignored when loading velocyto loomfiles.\\n\") }\n",
    "    return(X)\n",
    "  } else if (extension==\"csv\"){\n",
    "    X = read.csv(to_read, header = T, row.names = 1, sep = \",\", check.names = F) %>% as.matrix %>% Matrix(sparse = T)\n",
    "  } else {\n",
    "    assertthat::assert_that(file_test(\"-d\", to_read))\n",
    "    X = Read10X(to_read, version = 3)\n",
    "  }\n",
    "    # print(dim(X))\n",
    "  X = X[,Matrix::colSums(X) > min_umi_count]\n",
    "    # print(dim(X))\n",
    "  X\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eba6159f",
   "metadata": {},
   "source": [
    "- run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "25cb3bf4",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/data/pharynx_atlas/full_metadata.csv\"\n",
      "$CR_pharynx_E10_5_rep5\n",
      "[1] \"CR_pharynx_E10_5_rep5\"\n",
      "\n",
      "$CR_pharynx_E10_5_rep6\n",
      "[1] \"CR_pharynx_E10_5_rep6\"\n",
      "\n",
      "$CR_pharynx_E10_5_rep7\n",
      "[1] \"CR_pharynx_E10_5_rep7\"\n",
      "\n",
      "$CR_pharynx_E11_5_rep2\n",
      "[1] \"CR_pharynx_E11_5_rep2\"\n",
      "\n",
      "$CR_pharynx_E11_5_rep2B\n",
      "[1] \"CR_pharynx_E11_5_rep2B\"\n",
      "\n",
      "$CR_pharynx_E11_5_rep3\n",
      "[1] \"CR_pharynx_E11_5_rep3\"\n",
      "\n",
      "$CR_pharynx_E12_5_rep1\n",
      "[1] \"CR_pharynx_E12_5_rep1\"\n",
      "\n",
      "$CR_pharynx_E12_5_rep2\n",
      "[1] \"CR_pharynx_E12_5_rep2\"\n",
      "\n",
      "$CR_pharynx_E9_5_rep1\n",
      "[1] \"CR_pharynx_E9_5_rep1\"\n",
      "\n",
      "$CR_pharynx_E9_5_rep2\n",
      "[1] \"CR_pharynx_E9_5_rep2\"\n",
      "\n",
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/data/pharynx_atlas/metacell/CR_pharynx_E10_5_rep5\"\n",
      " [1] \"pharynx_atlas__CR_pharynx_E10_5_rep5__1\" \n",
      " [2] \"pharynx_atlas__CR_pharynx_E10_5_rep5__10\"\n",
      " [3] \"pharynx_atlas__CR_pharynx_E10_5_rep5__11\"\n",
      " [4] \"pharynx_atlas__CR_pharynx_E10_5_rep5__12\"\n",
      " [5] \"pharynx_atlas__CR_pharynx_E10_5_rep5__13\"\n",
      " [6] \"pharynx_atlas__CR_pharynx_E10_5_rep5__14\"\n",
      " [7] \"pharynx_atlas__CR_pharynx_E10_5_rep5__15\"\n",
      " [8] \"pharynx_atlas__CR_pharynx_E10_5_rep5__16\"\n",
      " [9] \"pharynx_atlas__CR_pharynx_E10_5_rep5__17\"\n",
      "[10] \"pharynx_atlas__CR_pharynx_E10_5_rep5__18\"\n",
      "[11] \"pharynx_atlas__CR_pharynx_E10_5_rep5__19\"\n",
      "[12] \"pharynx_atlas__CR_pharynx_E10_5_rep5__2\" \n",
      "[13] \"pharynx_atlas__CR_pharynx_E10_5_rep5__20\"\n",
      "[14] \"pharynx_atlas__CR_pharynx_E10_5_rep5__21\"\n",
      "[15] \"pharynx_atlas__CR_pharynx_E10_5_rep5__22\"\n",
      "[16] \"pharynx_atlas__CR_pharynx_E10_5_rep5__23\"\n",
      "[17] \"pharynx_atlas__CR_pharynx_E10_5_rep5__24\"\n",
      "[18] \"pharynx_atlas__CR_pharynx_E10_5_rep5__25\"\n",
      "[19] \"pharynx_atlas__CR_pharynx_E10_5_rep5__26\"\n",
      "[20] \"pharynx_atlas__CR_pharynx_E10_5_rep5__27\"\n",
      "[21] \"pharynx_atlas__CR_pharynx_E10_5_rep5__28\"\n",
      "[22] \"pharynx_atlas__CR_pharynx_E10_5_rep5__29\"\n",
      "[23] \"pharynx_atlas__CR_pharynx_E10_5_rep5__3\" \n",
      "[24] \"pharynx_atlas__CR_pharynx_E10_5_rep5__30\"\n",
      "[25] \"pharynx_atlas__CR_pharynx_E10_5_rep5__31\"\n",
      "[26] \"pharynx_atlas__CR_pharynx_E10_5_rep5__32\"\n",
      "[27] \"pharynx_atlas__CR_pharynx_E10_5_rep5__33\"\n",
      "[28] \"pharynx_atlas__CR_pharynx_E10_5_rep5__34\"\n",
      "[29] \"pharynx_atlas__CR_pharynx_E10_5_rep5__35\"\n",
      "[30] \"pharynx_atlas__CR_pharynx_E10_5_rep5__36\"\n",
      "[31] \"pharynx_atlas__CR_pharynx_E10_5_rep5__37\"\n",
      "[32] \"pharynx_atlas__CR_pharynx_E10_5_rep5__38\"\n",
      "[33] \"pharynx_atlas__CR_pharynx_E10_5_rep5__39\"\n",
      "[34] \"pharynx_atlas__CR_pharynx_E10_5_rep5__4\" \n",
      "[35] \"pharynx_atlas__CR_pharynx_E10_5_rep5__40\"\n",
      "[36] \"pharynx_atlas__CR_pharynx_E10_5_rep5__41\"\n",
      "[37] \"pharynx_atlas__CR_pharynx_E10_5_rep5__42\"\n",
      "[38] \"pharynx_atlas__CR_pharynx_E10_5_rep5__43\"\n",
      "[39] \"pharynx_atlas__CR_pharynx_E10_5_rep5__44\"\n",
      "[40] \"pharynx_atlas__CR_pharynx_E10_5_rep5__45\"\n",
      "[41] \"pharynx_atlas__CR_pharynx_E10_5_rep5__46\"\n",
      "[42] \"pharynx_atlas__CR_pharynx_E10_5_rep5__47\"\n",
      "[43] \"pharynx_atlas__CR_pharynx_E10_5_rep5__48\"\n",
      "[44] \"pharynx_atlas__CR_pharynx_E10_5_rep5__49\"\n",
      "[45] \"pharynx_atlas__CR_pharynx_E10_5_rep5__5\" \n",
      "[46] \"pharynx_atlas__CR_pharynx_E10_5_rep5__50\"\n",
      "[47] \"pharynx_atlas__CR_pharynx_E10_5_rep5__51\"\n",
      "[48] \"pharynx_atlas__CR_pharynx_E10_5_rep5__52\"\n",
      "[49] \"pharynx_atlas__CR_pharynx_E10_5_rep5__53\"\n",
      "[50] \"pharynx_atlas__CR_pharynx_E10_5_rep5__6\" \n",
      "[51] \"pharynx_atlas__CR_pharynx_E10_5_rep5__7\" \n",
      "[52] \"pharynx_atlas__CR_pharynx_E10_5_rep5__8\" \n",
      "[53] \"pharynx_atlas__CR_pharynx_E10_5_rep5__9\" \n",
      "[1] 27998    53\n",
      "[1] 27998    53\n",
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/data/pharynx_atlas/metacell/CR_pharynx_E10_5_rep6\"\n",
      " [1] \"pharynx_atlas__CR_pharynx_E10_5_rep6__1\" \n",
      " [2] \"pharynx_atlas__CR_pharynx_E10_5_rep6__10\"\n",
      " [3] \"pharynx_atlas__CR_pharynx_E10_5_rep6__11\"\n",
      " [4] \"pharynx_atlas__CR_pharynx_E10_5_rep6__12\"\n",
      " [5] \"pharynx_atlas__CR_pharynx_E10_5_rep6__13\"\n",
      " [6] \"pharynx_atlas__CR_pharynx_E10_5_rep6__14\"\n",
      " [7] \"pharynx_atlas__CR_pharynx_E10_5_rep6__15\"\n",
      " [8] \"pharynx_atlas__CR_pharynx_E10_5_rep6__16\"\n",
      " [9] \"pharynx_atlas__CR_pharynx_E10_5_rep6__17\"\n",
      "[10] \"pharynx_atlas__CR_pharynx_E10_5_rep6__18\"\n",
      "[11] \"pharynx_atlas__CR_pharynx_E10_5_rep6__19\"\n",
      "[12] \"pharynx_atlas__CR_pharynx_E10_5_rep6__2\" \n",
      "[13] \"pharynx_atlas__CR_pharynx_E10_5_rep6__20\"\n",
      "[14] \"pharynx_atlas__CR_pharynx_E10_5_rep6__21\"\n",
      "[15] \"pharynx_atlas__CR_pharynx_E10_5_rep6__22\"\n",
      "[16] \"pharynx_atlas__CR_pharynx_E10_5_rep6__23\"\n",
      "[17] \"pharynx_atlas__CR_pharynx_E10_5_rep6__24\"\n",
      "[18] \"pharynx_atlas__CR_pharynx_E10_5_rep6__25\"\n",
      "[19] \"pharynx_atlas__CR_pharynx_E10_5_rep6__26\"\n",
      "[20] \"pharynx_atlas__CR_pharynx_E10_5_rep6__27\"\n",
      "[21] \"pharynx_atlas__CR_pharynx_E10_5_rep6__28\"\n",
      "[22] \"pharynx_atlas__CR_pharynx_E10_5_rep6__29\"\n",
      "[23] \"pharynx_atlas__CR_pharynx_E10_5_rep6__3\" \n",
      "[24] \"pharynx_atlas__CR_pharynx_E10_5_rep6__30\"\n",
      "[25] \"pharynx_atlas__CR_pharynx_E10_5_rep6__31\"\n",
      "[26] \"pharynx_atlas__CR_pharynx_E10_5_rep6__32\"\n",
      "[27] \"pharynx_atlas__CR_pharynx_E10_5_rep6__33\"\n",
      "[28] \"pharynx_atlas__CR_pharynx_E10_5_rep6__34\"\n",
      "[29] \"pharynx_atlas__CR_pharynx_E10_5_rep6__35\"\n",
      "[30] \"pharynx_atlas__CR_pharynx_E10_5_rep6__36\"\n",
      "[31] \"pharynx_atlas__CR_pharynx_E10_5_rep6__37\"\n",
      "[32] \"pharynx_atlas__CR_pharynx_E10_5_rep6__38\"\n",
      "[33] \"pharynx_atlas__CR_pharynx_E10_5_rep6__39\"\n",
      "[34] \"pharynx_atlas__CR_pharynx_E10_5_rep6__4\" \n",
      "[35] \"pharynx_atlas__CR_pharynx_E10_5_rep6__40\"\n",
      "[36] \"pharynx_atlas__CR_pharynx_E10_5_rep6__41\"\n",
      "[37] \"pharynx_atlas__CR_pharynx_E10_5_rep6__42\"\n",
      "[38] \"pharynx_atlas__CR_pharynx_E10_5_rep6__43\"\n",
      "[39] \"pharynx_atlas__CR_pharynx_E10_5_rep6__44\"\n",
      "[40] \"pharynx_atlas__CR_pharynx_E10_5_rep6__45\"\n",
      "[41] \"pharynx_atlas__CR_pharynx_E10_5_rep6__46\"\n",
      "[42] \"pharynx_atlas__CR_pharynx_E10_5_rep6__5\" \n",
      "[43] \"pharynx_atlas__CR_pharynx_E10_5_rep6__6\" \n",
      "[44] \"pharynx_atlas__CR_pharynx_E10_5_rep6__7\" \n",
      "[45] \"pharynx_atlas__CR_pharynx_E10_5_rep6__8\" \n",
      "[46] \"pharynx_atlas__CR_pharynx_E10_5_rep6__9\" \n",
      "[1] 27998    46\n",
      "[1] 27998    46\n",
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/data/pharynx_atlas/metacell/CR_pharynx_E10_5_rep7\"\n",
      " [1] \"pharynx_atlas__CR_pharynx_E10_5_rep7__1\" \n",
      " [2] \"pharynx_atlas__CR_pharynx_E10_5_rep7__10\"\n",
      " [3] \"pharynx_atlas__CR_pharynx_E10_5_rep7__11\"\n",
      " [4] \"pharynx_atlas__CR_pharynx_E10_5_rep7__12\"\n",
      " [5] \"pharynx_atlas__CR_pharynx_E10_5_rep7__13\"\n",
      " [6] \"pharynx_atlas__CR_pharynx_E10_5_rep7__14\"\n",
      " [7] \"pharynx_atlas__CR_pharynx_E10_5_rep7__15\"\n",
      " [8] \"pharynx_atlas__CR_pharynx_E10_5_rep7__16\"\n",
      " [9] \"pharynx_atlas__CR_pharynx_E10_5_rep7__17\"\n",
      "[10] \"pharynx_atlas__CR_pharynx_E10_5_rep7__18\"\n",
      "[11] \"pharynx_atlas__CR_pharynx_E10_5_rep7__19\"\n",
      "[12] \"pharynx_atlas__CR_pharynx_E10_5_rep7__2\" \n",
      "[13] \"pharynx_atlas__CR_pharynx_E10_5_rep7__20\"\n",
      "[14] \"pharynx_atlas__CR_pharynx_E10_5_rep7__21\"\n",
      "[15] \"pharynx_atlas__CR_pharynx_E10_5_rep7__22\"\n",
      "[16] \"pharynx_atlas__CR_pharynx_E10_5_rep7__23\"\n",
      "[17] \"pharynx_atlas__CR_pharynx_E10_5_rep7__24\"\n",
      "[18] \"pharynx_atlas__CR_pharynx_E10_5_rep7__25\"\n",
      "[19] \"pharynx_atlas__CR_pharynx_E10_5_rep7__26\"\n",
      "[20] \"pharynx_atlas__CR_pharynx_E10_5_rep7__27\"\n",
      "[21] \"pharynx_atlas__CR_pharynx_E10_5_rep7__28\"\n",
      "[22] \"pharynx_atlas__CR_pharynx_E10_5_rep7__29\"\n",
      "[23] \"pharynx_atlas__CR_pharynx_E10_5_rep7__3\" \n",
      "[24] \"pharynx_atlas__CR_pharynx_E10_5_rep7__30\"\n",
      "[25] \"pharynx_atlas__CR_pharynx_E10_5_rep7__31\"\n",
      "[26] \"pharynx_atlas__CR_pharynx_E10_5_rep7__32\"\n",
      "[27] \"pharynx_atlas__CR_pharynx_E10_5_rep7__33\"\n",
      "[28] \"pharynx_atlas__CR_pharynx_E10_5_rep7__34\"\n",
      "[29] \"pharynx_atlas__CR_pharynx_E10_5_rep7__35\"\n",
      "[30] \"pharynx_atlas__CR_pharynx_E10_5_rep7__36\"\n",
      "[31] \"pharynx_atlas__CR_pharynx_E10_5_rep7__37\"\n",
      "[32] \"pharynx_atlas__CR_pharynx_E10_5_rep7__38\"\n",
      "[33] \"pharynx_atlas__CR_pharynx_E10_5_rep7__39\"\n",
      "[34] \"pharynx_atlas__CR_pharynx_E10_5_rep7__4\" \n",
      "[35] \"pharynx_atlas__CR_pharynx_E10_5_rep7__40\"\n",
      "[36] \"pharynx_atlas__CR_pharynx_E10_5_rep7__41\"\n",
      "[37] \"pharynx_atlas__CR_pharynx_E10_5_rep7__42\"\n",
      "[38] \"pharynx_atlas__CR_pharynx_E10_5_rep7__43\"\n",
      "[39] \"pharynx_atlas__CR_pharynx_E10_5_rep7__44\"\n",
      "[40] \"pharynx_atlas__CR_pharynx_E10_5_rep7__45\"\n",
      "[41] \"pharynx_atlas__CR_pharynx_E10_5_rep7__46\"\n",
      "[42] \"pharynx_atlas__CR_pharynx_E10_5_rep7__47\"\n",
      "[43] \"pharynx_atlas__CR_pharynx_E10_5_rep7__5\" \n",
      "[44] \"pharynx_atlas__CR_pharynx_E10_5_rep7__6\" \n",
      "[45] \"pharynx_atlas__CR_pharynx_E10_5_rep7__7\" \n",
      "[46] \"pharynx_atlas__CR_pharynx_E10_5_rep7__8\" \n",
      "[47] \"pharynx_atlas__CR_pharynx_E10_5_rep7__9\" \n",
      "[1] 27998    47\n",
      "[1] 27998    47\n",
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/data/pharynx_atlas/metacell/CR_pharynx_E11_5_rep2\"\n",
      " [1] \"pharynx_atlas__CR_pharynx_E11_5_rep2__1\" \n",
      " [2] \"pharynx_atlas__CR_pharynx_E11_5_rep2__10\"\n",
      " [3] \"pharynx_atlas__CR_pharynx_E11_5_rep2__11\"\n",
      " [4] \"pharynx_atlas__CR_pharynx_E11_5_rep2__12\"\n",
      " [5] \"pharynx_atlas__CR_pharynx_E11_5_rep2__13\"\n",
      " [6] \"pharynx_atlas__CR_pharynx_E11_5_rep2__14\"\n",
      " [7] \"pharynx_atlas__CR_pharynx_E11_5_rep2__15\"\n",
      " [8] \"pharynx_atlas__CR_pharynx_E11_5_rep2__16\"\n",
      " [9] \"pharynx_atlas__CR_pharynx_E11_5_rep2__17\"\n",
      "[10] \"pharynx_atlas__CR_pharynx_E11_5_rep2__18\"\n",
      "[11] \"pharynx_atlas__CR_pharynx_E11_5_rep2__19\"\n",
      "[12] \"pharynx_atlas__CR_pharynx_E11_5_rep2__2\" \n",
      "[13] \"pharynx_atlas__CR_pharynx_E11_5_rep2__20\"\n",
      "[14] \"pharynx_atlas__CR_pharynx_E11_5_rep2__21\"\n",
      "[15] \"pharynx_atlas__CR_pharynx_E11_5_rep2__22\"\n",
      "[16] \"pharynx_atlas__CR_pharynx_E11_5_rep2__23\"\n",
      "[17] \"pharynx_atlas__CR_pharynx_E11_5_rep2__24\"\n",
      "[18] \"pharynx_atlas__CR_pharynx_E11_5_rep2__25\"\n",
      "[19] \"pharynx_atlas__CR_pharynx_E11_5_rep2__26\"\n",
      "[20] \"pharynx_atlas__CR_pharynx_E11_5_rep2__27\"\n",
      "[21] \"pharynx_atlas__CR_pharynx_E11_5_rep2__28\"\n",
      "[22] \"pharynx_atlas__CR_pharynx_E11_5_rep2__29\"\n",
      "[23] \"pharynx_atlas__CR_pharynx_E11_5_rep2__3\" \n",
      "[24] \"pharynx_atlas__CR_pharynx_E11_5_rep2__30\"\n",
      "[25] \"pharynx_atlas__CR_pharynx_E11_5_rep2__31\"\n",
      "[26] \"pharynx_atlas__CR_pharynx_E11_5_rep2__32\"\n",
      "[27] \"pharynx_atlas__CR_pharynx_E11_5_rep2__33\"\n",
      "[28] \"pharynx_atlas__CR_pharynx_E11_5_rep2__34\"\n",
      "[29] \"pharynx_atlas__CR_pharynx_E11_5_rep2__35\"\n",
      "[30] \"pharynx_atlas__CR_pharynx_E11_5_rep2__36\"\n",
      "[31] \"pharynx_atlas__CR_pharynx_E11_5_rep2__37\"\n",
      "[32] \"pharynx_atlas__CR_pharynx_E11_5_rep2__38\"\n",
      "[33] \"pharynx_atlas__CR_pharynx_E11_5_rep2__39\"\n",
      "[34] \"pharynx_atlas__CR_pharynx_E11_5_rep2__4\" \n",
      "[35] \"pharynx_atlas__CR_pharynx_E11_5_rep2__40\"\n",
      "[36] \"pharynx_atlas__CR_pharynx_E11_5_rep2__41\"\n",
      "[37] \"pharynx_atlas__CR_pharynx_E11_5_rep2__42\"\n",
      "[38] \"pharynx_atlas__CR_pharynx_E11_5_rep2__43\"\n",
      "[39] \"pharynx_atlas__CR_pharynx_E11_5_rep2__44\"\n",
      "[40] \"pharynx_atlas__CR_pharynx_E11_5_rep2__45\"\n",
      "[41] \"pharynx_atlas__CR_pharynx_E11_5_rep2__46\"\n",
      "[42] \"pharynx_atlas__CR_pharynx_E11_5_rep2__47\"\n",
      "[43] \"pharynx_atlas__CR_pharynx_E11_5_rep2__48\"\n",
      "[44] \"pharynx_atlas__CR_pharynx_E11_5_rep2__49\"\n",
      "[45] \"pharynx_atlas__CR_pharynx_E11_5_rep2__5\" \n",
      "[46] \"pharynx_atlas__CR_pharynx_E11_5_rep2__50\"\n",
      "[47] \"pharynx_atlas__CR_pharynx_E11_5_rep2__51\"\n",
      "[48] \"pharynx_atlas__CR_pharynx_E11_5_rep2__52\"\n",
      "[49] \"pharynx_atlas__CR_pharynx_E11_5_rep2__53\"\n",
      "[50] \"pharynx_atlas__CR_pharynx_E11_5_rep2__54\"\n",
      "[51] \"pharynx_atlas__CR_pharynx_E11_5_rep2__55\"\n",
      "[52] \"pharynx_atlas__CR_pharynx_E11_5_rep2__56\"\n",
      "[53] \"pharynx_atlas__CR_pharynx_E11_5_rep2__57\"\n",
      "[54] \"pharynx_atlas__CR_pharynx_E11_5_rep2__58\"\n",
      "[55] \"pharynx_atlas__CR_pharynx_E11_5_rep2__6\" \n",
      "[56] \"pharynx_atlas__CR_pharynx_E11_5_rep2__7\" \n",
      "[57] \"pharynx_atlas__CR_pharynx_E11_5_rep2__8\" \n",
      "[58] \"pharynx_atlas__CR_pharynx_E11_5_rep2__9\" \n",
      "[1] 27998    58\n",
      "[1] 27998    58\n",
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/data/pharynx_atlas/metacell/CR_pharynx_E11_5_rep2B\"\n",
      " [1] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__1\" \n",
      " [2] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__10\"\n",
      " [3] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__11\"\n",
      " [4] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__12\"\n",
      " [5] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__13\"\n",
      " [6] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__14\"\n",
      " [7] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__15\"\n",
      " [8] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__16\"\n",
      " [9] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__17\"\n",
      "[10] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__18\"\n",
      "[11] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__19\"\n",
      "[12] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__2\" \n",
      "[13] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__20\"\n",
      "[14] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__21\"\n",
      "[15] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__22\"\n",
      "[16] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__23\"\n",
      "[17] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__24\"\n",
      "[18] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__25\"\n",
      "[19] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__26\"\n",
      "[20] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__27\"\n",
      "[21] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__28\"\n",
      "[22] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__29\"\n",
      "[23] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__3\" \n",
      "[24] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__30\"\n",
      "[25] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__31\"\n",
      "[26] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__32\"\n",
      "[27] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__33\"\n",
      "[28] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__34\"\n",
      "[29] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__35\"\n",
      "[30] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__4\" \n",
      "[31] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__5\" \n",
      "[32] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__6\" \n",
      "[33] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__7\" \n",
      "[34] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__8\" \n",
      "[35] \"pharynx_atlas__CR_pharynx_E11_5_rep2B__9\" \n",
      "[1] 27998    35\n",
      "[1] 27998    35\n",
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/data/pharynx_atlas/metacell/CR_pharynx_E11_5_rep3\"\n",
      " [1] \"pharynx_atlas__CR_pharynx_E11_5_rep3__1\" \n",
      " [2] \"pharynx_atlas__CR_pharynx_E11_5_rep3__10\"\n",
      " [3] \"pharynx_atlas__CR_pharynx_E11_5_rep3__11\"\n",
      " [4] \"pharynx_atlas__CR_pharynx_E11_5_rep3__12\"\n",
      " [5] \"pharynx_atlas__CR_pharynx_E11_5_rep3__13\"\n",
      " [6] \"pharynx_atlas__CR_pharynx_E11_5_rep3__14\"\n",
      " [7] \"pharynx_atlas__CR_pharynx_E11_5_rep3__15\"\n",
      " [8] \"pharynx_atlas__CR_pharynx_E11_5_rep3__16\"\n",
      " [9] \"pharynx_atlas__CR_pharynx_E11_5_rep3__17\"\n",
      "[10] \"pharynx_atlas__CR_pharynx_E11_5_rep3__18\"\n",
      "[11] \"pharynx_atlas__CR_pharynx_E11_5_rep3__19\"\n",
      "[12] \"pharynx_atlas__CR_pharynx_E11_5_rep3__2\" \n",
      "[13] \"pharynx_atlas__CR_pharynx_E11_5_rep3__20\"\n",
      "[14] \"pharynx_atlas__CR_pharynx_E11_5_rep3__21\"\n",
      "[15] \"pharynx_atlas__CR_pharynx_E11_5_rep3__22\"\n",
      "[16] \"pharynx_atlas__CR_pharynx_E11_5_rep3__23\"\n",
      "[17] \"pharynx_atlas__CR_pharynx_E11_5_rep3__24\"\n",
      "[18] \"pharynx_atlas__CR_pharynx_E11_5_rep3__25\"\n",
      "[19] \"pharynx_atlas__CR_pharynx_E11_5_rep3__26\"\n",
      "[20] \"pharynx_atlas__CR_pharynx_E11_5_rep3__27\"\n",
      "[21] \"pharynx_atlas__CR_pharynx_E11_5_rep3__28\"\n",
      "[22] \"pharynx_atlas__CR_pharynx_E11_5_rep3__29\"\n",
      "[23] \"pharynx_atlas__CR_pharynx_E11_5_rep3__3\" \n",
      "[24] \"pharynx_atlas__CR_pharynx_E11_5_rep3__30\"\n",
      "[25] \"pharynx_atlas__CR_pharynx_E11_5_rep3__31\"\n",
      "[26] \"pharynx_atlas__CR_pharynx_E11_5_rep3__32\"\n",
      "[27] \"pharynx_atlas__CR_pharynx_E11_5_rep3__33\"\n",
      "[28] \"pharynx_atlas__CR_pharynx_E11_5_rep3__34\"\n",
      "[29] \"pharynx_atlas__CR_pharynx_E11_5_rep3__35\"\n",
      "[30] \"pharynx_atlas__CR_pharynx_E11_5_rep3__36\"\n",
      "[31] \"pharynx_atlas__CR_pharynx_E11_5_rep3__37\"\n",
      "[32] \"pharynx_atlas__CR_pharynx_E11_5_rep3__38\"\n",
      "[33] \"pharynx_atlas__CR_pharynx_E11_5_rep3__39\"\n",
      "[34] \"pharynx_atlas__CR_pharynx_E11_5_rep3__4\" \n",
      "[35] \"pharynx_atlas__CR_pharynx_E11_5_rep3__40\"\n",
      "[36] \"pharynx_atlas__CR_pharynx_E11_5_rep3__41\"\n",
      "[37] \"pharynx_atlas__CR_pharynx_E11_5_rep3__42\"\n",
      "[38] \"pharynx_atlas__CR_pharynx_E11_5_rep3__43\"\n",
      "[39] \"pharynx_atlas__CR_pharynx_E11_5_rep3__44\"\n",
      "[40] \"pharynx_atlas__CR_pharynx_E11_5_rep3__45\"\n",
      "[41] \"pharynx_atlas__CR_pharynx_E11_5_rep3__46\"\n",
      "[42] \"pharynx_atlas__CR_pharynx_E11_5_rep3__47\"\n",
      "[43] \"pharynx_atlas__CR_pharynx_E11_5_rep3__48\"\n",
      "[44] \"pharynx_atlas__CR_pharynx_E11_5_rep3__49\"\n",
      "[45] \"pharynx_atlas__CR_pharynx_E11_5_rep3__5\" \n",
      "[46] \"pharynx_atlas__CR_pharynx_E11_5_rep3__50\"\n",
      "[47] \"pharynx_atlas__CR_pharynx_E11_5_rep3__51\"\n",
      "[48] \"pharynx_atlas__CR_pharynx_E11_5_rep3__52\"\n",
      "[49] \"pharynx_atlas__CR_pharynx_E11_5_rep3__53\"\n",
      "[50] \"pharynx_atlas__CR_pharynx_E11_5_rep3__54\"\n",
      "[51] \"pharynx_atlas__CR_pharynx_E11_5_rep3__55\"\n",
      "[52] \"pharynx_atlas__CR_pharynx_E11_5_rep3__56\"\n",
      "[53] \"pharynx_atlas__CR_pharynx_E11_5_rep3__57\"\n",
      "[54] \"pharynx_atlas__CR_pharynx_E11_5_rep3__58\"\n",
      "[55] \"pharynx_atlas__CR_pharynx_E11_5_rep3__59\"\n",
      "[56] \"pharynx_atlas__CR_pharynx_E11_5_rep3__6\" \n",
      "[57] \"pharynx_atlas__CR_pharynx_E11_5_rep3__60\"\n",
      "[58] \"pharynx_atlas__CR_pharynx_E11_5_rep3__61\"\n",
      "[59] \"pharynx_atlas__CR_pharynx_E11_5_rep3__62\"\n",
      "[60] \"pharynx_atlas__CR_pharynx_E11_5_rep3__63\"\n",
      "[61] \"pharynx_atlas__CR_pharynx_E11_5_rep3__64\"\n",
      "[62] \"pharynx_atlas__CR_pharynx_E11_5_rep3__65\"\n",
      "[63] \"pharynx_atlas__CR_pharynx_E11_5_rep3__66\"\n",
      "[64] \"pharynx_atlas__CR_pharynx_E11_5_rep3__67\"\n",
      "[65] \"pharynx_atlas__CR_pharynx_E11_5_rep3__68\"\n",
      "[66] \"pharynx_atlas__CR_pharynx_E11_5_rep3__69\"\n",
      "[67] \"pharynx_atlas__CR_pharynx_E11_5_rep3__7\" \n",
      "[68] \"pharynx_atlas__CR_pharynx_E11_5_rep3__70\"\n",
      "[69] \"pharynx_atlas__CR_pharynx_E11_5_rep3__71\"\n",
      "[70] \"pharynx_atlas__CR_pharynx_E11_5_rep3__8\" \n",
      "[71] \"pharynx_atlas__CR_pharynx_E11_5_rep3__9\" \n",
      "[1] 27998    71\n",
      "[1] 27998    71\n",
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/data/pharynx_atlas/metacell/CR_pharynx_E12_5_rep1\"\n",
      " [1] \"pharynx_atlas__CR_pharynx_E12_5_rep1__1\" \n",
      " [2] \"pharynx_atlas__CR_pharynx_E12_5_rep1__10\"\n",
      " [3] \"pharynx_atlas__CR_pharynx_E12_5_rep1__11\"\n",
      " [4] \"pharynx_atlas__CR_pharynx_E12_5_rep1__12\"\n",
      " [5] \"pharynx_atlas__CR_pharynx_E12_5_rep1__13\"\n",
      " [6] \"pharynx_atlas__CR_pharynx_E12_5_rep1__14\"\n",
      " [7] \"pharynx_atlas__CR_pharynx_E12_5_rep1__15\"\n",
      " [8] \"pharynx_atlas__CR_pharynx_E12_5_rep1__16\"\n",
      " [9] \"pharynx_atlas__CR_pharynx_E12_5_rep1__17\"\n",
      "[10] \"pharynx_atlas__CR_pharynx_E12_5_rep1__18\"\n",
      "[11] \"pharynx_atlas__CR_pharynx_E12_5_rep1__19\"\n",
      "[12] \"pharynx_atlas__CR_pharynx_E12_5_rep1__2\" \n",
      "[13] \"pharynx_atlas__CR_pharynx_E12_5_rep1__20\"\n",
      "[14] \"pharynx_atlas__CR_pharynx_E12_5_rep1__21\"\n",
      "[15] \"pharynx_atlas__CR_pharynx_E12_5_rep1__22\"\n",
      "[16] \"pharynx_atlas__CR_pharynx_E12_5_rep1__23\"\n",
      "[17] \"pharynx_atlas__CR_pharynx_E12_5_rep1__24\"\n",
      "[18] \"pharynx_atlas__CR_pharynx_E12_5_rep1__25\"\n",
      "[19] \"pharynx_atlas__CR_pharynx_E12_5_rep1__26\"\n",
      "[20] \"pharynx_atlas__CR_pharynx_E12_5_rep1__27\"\n",
      "[21] \"pharynx_atlas__CR_pharynx_E12_5_rep1__28\"\n",
      "[22] \"pharynx_atlas__CR_pharynx_E12_5_rep1__29\"\n",
      "[23] \"pharynx_atlas__CR_pharynx_E12_5_rep1__3\" \n",
      "[24] \"pharynx_atlas__CR_pharynx_E12_5_rep1__30\"\n",
      "[25] \"pharynx_atlas__CR_pharynx_E12_5_rep1__31\"\n",
      "[26] \"pharynx_atlas__CR_pharynx_E12_5_rep1__32\"\n",
      "[27] \"pharynx_atlas__CR_pharynx_E12_5_rep1__33\"\n",
      "[28] \"pharynx_atlas__CR_pharynx_E12_5_rep1__34\"\n",
      "[29] \"pharynx_atlas__CR_pharynx_E12_5_rep1__35\"\n",
      "[30] \"pharynx_atlas__CR_pharynx_E12_5_rep1__36\"\n",
      "[31] \"pharynx_atlas__CR_pharynx_E12_5_rep1__37\"\n",
      "[32] \"pharynx_atlas__CR_pharynx_E12_5_rep1__38\"\n",
      "[33] \"pharynx_atlas__CR_pharynx_E12_5_rep1__39\"\n",
      "[34] \"pharynx_atlas__CR_pharynx_E12_5_rep1__4\" \n",
      "[35] \"pharynx_atlas__CR_pharynx_E12_5_rep1__40\"\n",
      "[36] \"pharynx_atlas__CR_pharynx_E12_5_rep1__41\"\n",
      "[37] \"pharynx_atlas__CR_pharynx_E12_5_rep1__42\"\n",
      "[38] \"pharynx_atlas__CR_pharynx_E12_5_rep1__43\"\n",
      "[39] \"pharynx_atlas__CR_pharynx_E12_5_rep1__44\"\n",
      "[40] \"pharynx_atlas__CR_pharynx_E12_5_rep1__45\"\n",
      "[41] \"pharynx_atlas__CR_pharynx_E12_5_rep1__46\"\n",
      "[42] \"pharynx_atlas__CR_pharynx_E12_5_rep1__47\"\n",
      "[43] \"pharynx_atlas__CR_pharynx_E12_5_rep1__48\"\n",
      "[44] \"pharynx_atlas__CR_pharynx_E12_5_rep1__49\"\n",
      "[45] \"pharynx_atlas__CR_pharynx_E12_5_rep1__5\" \n",
      "[46] \"pharynx_atlas__CR_pharynx_E12_5_rep1__50\"\n",
      "[47] \"pharynx_atlas__CR_pharynx_E12_5_rep1__6\" \n",
      "[48] \"pharynx_atlas__CR_pharynx_E12_5_rep1__7\" \n",
      "[49] \"pharynx_atlas__CR_pharynx_E12_5_rep1__8\" \n",
      "[50] \"pharynx_atlas__CR_pharynx_E12_5_rep1__9\" \n",
      "[1] 27998    50\n",
      "[1] 27998    50\n",
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/data/pharynx_atlas/metacell/CR_pharynx_E12_5_rep2\"\n",
      " [1] \"pharynx_atlas__CR_pharynx_E12_5_rep2__1\" \n",
      " [2] \"pharynx_atlas__CR_pharynx_E12_5_rep2__10\"\n",
      " [3] \"pharynx_atlas__CR_pharynx_E12_5_rep2__11\"\n",
      " [4] \"pharynx_atlas__CR_pharynx_E12_5_rep2__12\"\n",
      " [5] \"pharynx_atlas__CR_pharynx_E12_5_rep2__13\"\n",
      " [6] \"pharynx_atlas__CR_pharynx_E12_5_rep2__14\"\n",
      " [7] \"pharynx_atlas__CR_pharynx_E12_5_rep2__15\"\n",
      " [8] \"pharynx_atlas__CR_pharynx_E12_5_rep2__16\"\n",
      " [9] \"pharynx_atlas__CR_pharynx_E12_5_rep2__17\"\n",
      "[10] \"pharynx_atlas__CR_pharynx_E12_5_rep2__18\"\n",
      "[11] \"pharynx_atlas__CR_pharynx_E12_5_rep2__19\"\n",
      "[12] \"pharynx_atlas__CR_pharynx_E12_5_rep2__2\" \n",
      "[13] \"pharynx_atlas__CR_pharynx_E12_5_rep2__20\"\n",
      "[14] \"pharynx_atlas__CR_pharynx_E12_5_rep2__21\"\n",
      "[15] \"pharynx_atlas__CR_pharynx_E12_5_rep2__22\"\n",
      "[16] \"pharynx_atlas__CR_pharynx_E12_5_rep2__23\"\n",
      "[17] \"pharynx_atlas__CR_pharynx_E12_5_rep2__24\"\n",
      "[18] \"pharynx_atlas__CR_pharynx_E12_5_rep2__25\"\n",
      "[19] \"pharynx_atlas__CR_pharynx_E12_5_rep2__26\"\n",
      "[20] \"pharynx_atlas__CR_pharynx_E12_5_rep2__27\"\n",
      "[21] \"pharynx_atlas__CR_pharynx_E12_5_rep2__28\"\n",
      "[22] \"pharynx_atlas__CR_pharynx_E12_5_rep2__29\"\n",
      "[23] \"pharynx_atlas__CR_pharynx_E12_5_rep2__3\" \n",
      "[24] \"pharynx_atlas__CR_pharynx_E12_5_rep2__30\"\n",
      "[25] \"pharynx_atlas__CR_pharynx_E12_5_rep2__31\"\n",
      "[26] \"pharynx_atlas__CR_pharynx_E12_5_rep2__32\"\n",
      "[27] \"pharynx_atlas__CR_pharynx_E12_5_rep2__33\"\n",
      "[28] \"pharynx_atlas__CR_pharynx_E12_5_rep2__34\"\n",
      "[29] \"pharynx_atlas__CR_pharynx_E12_5_rep2__35\"\n",
      "[30] \"pharynx_atlas__CR_pharynx_E12_5_rep2__36\"\n",
      "[31] \"pharynx_atlas__CR_pharynx_E12_5_rep2__37\"\n",
      "[32] \"pharynx_atlas__CR_pharynx_E12_5_rep2__38\"\n",
      "[33] \"pharynx_atlas__CR_pharynx_E12_5_rep2__39\"\n",
      "[34] \"pharynx_atlas__CR_pharynx_E12_5_rep2__4\" \n",
      "[35] \"pharynx_atlas__CR_pharynx_E12_5_rep2__40\"\n",
      "[36] \"pharynx_atlas__CR_pharynx_E12_5_rep2__41\"\n",
      "[37] \"pharynx_atlas__CR_pharynx_E12_5_rep2__42\"\n",
      "[38] \"pharynx_atlas__CR_pharynx_E12_5_rep2__43\"\n",
      "[39] \"pharynx_atlas__CR_pharynx_E12_5_rep2__44\"\n",
      "[40] \"pharynx_atlas__CR_pharynx_E12_5_rep2__45\"\n",
      "[41] \"pharynx_atlas__CR_pharynx_E12_5_rep2__46\"\n",
      "[42] \"pharynx_atlas__CR_pharynx_E12_5_rep2__47\"\n",
      "[43] \"pharynx_atlas__CR_pharynx_E12_5_rep2__48\"\n",
      "[44] \"pharynx_atlas__CR_pharynx_E12_5_rep2__49\"\n",
      "[45] \"pharynx_atlas__CR_pharynx_E12_5_rep2__5\" \n",
      "[46] \"pharynx_atlas__CR_pharynx_E12_5_rep2__50\"\n",
      "[47] \"pharynx_atlas__CR_pharynx_E12_5_rep2__51\"\n",
      "[48] \"pharynx_atlas__CR_pharynx_E12_5_rep2__52\"\n",
      "[49] \"pharynx_atlas__CR_pharynx_E12_5_rep2__53\"\n",
      "[50] \"pharynx_atlas__CR_pharynx_E12_5_rep2__54\"\n",
      "[51] \"pharynx_atlas__CR_pharynx_E12_5_rep2__55\"\n",
      "[52] \"pharynx_atlas__CR_pharynx_E12_5_rep2__56\"\n",
      "[53] \"pharynx_atlas__CR_pharynx_E12_5_rep2__57\"\n",
      "[54] \"pharynx_atlas__CR_pharynx_E12_5_rep2__58\"\n",
      "[55] \"pharynx_atlas__CR_pharynx_E12_5_rep2__59\"\n",
      "[56] \"pharynx_atlas__CR_pharynx_E12_5_rep2__6\" \n",
      "[57] \"pharynx_atlas__CR_pharynx_E12_5_rep2__60\"\n",
      "[58] \"pharynx_atlas__CR_pharynx_E12_5_rep2__61\"\n",
      "[59] \"pharynx_atlas__CR_pharynx_E12_5_rep2__62\"\n",
      "[60] \"pharynx_atlas__CR_pharynx_E12_5_rep2__63\"\n",
      "[61] \"pharynx_atlas__CR_pharynx_E12_5_rep2__64\"\n",
      "[62] \"pharynx_atlas__CR_pharynx_E12_5_rep2__7\" \n",
      "[63] \"pharynx_atlas__CR_pharynx_E12_5_rep2__8\" \n",
      "[64] \"pharynx_atlas__CR_pharynx_E12_5_rep2__9\" \n",
      "[1] 27998    64\n",
      "[1] 27998    64\n",
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/data/pharynx_atlas/metacell/CR_pharynx_E9_5_rep1\"\n",
      " [1] \"pharynx_atlas__CR_pharynx_E9_5_rep1__1\" \n",
      " [2] \"pharynx_atlas__CR_pharynx_E9_5_rep1__10\"\n",
      " [3] \"pharynx_atlas__CR_pharynx_E9_5_rep1__11\"\n",
      " [4] \"pharynx_atlas__CR_pharynx_E9_5_rep1__12\"\n",
      " [5] \"pharynx_atlas__CR_pharynx_E9_5_rep1__13\"\n",
      " [6] \"pharynx_atlas__CR_pharynx_E9_5_rep1__14\"\n",
      " [7] \"pharynx_atlas__CR_pharynx_E9_5_rep1__15\"\n",
      " [8] \"pharynx_atlas__CR_pharynx_E9_5_rep1__16\"\n",
      " [9] \"pharynx_atlas__CR_pharynx_E9_5_rep1__17\"\n",
      "[10] \"pharynx_atlas__CR_pharynx_E9_5_rep1__18\"\n",
      "[11] \"pharynx_atlas__CR_pharynx_E9_5_rep1__19\"\n",
      "[12] \"pharynx_atlas__CR_pharynx_E9_5_rep1__2\" \n",
      "[13] \"pharynx_atlas__CR_pharynx_E9_5_rep1__20\"\n",
      "[14] \"pharynx_atlas__CR_pharynx_E9_5_rep1__21\"\n",
      "[15] \"pharynx_atlas__CR_pharynx_E9_5_rep1__22\"\n",
      "[16] \"pharynx_atlas__CR_pharynx_E9_5_rep1__23\"\n",
      "[17] \"pharynx_atlas__CR_pharynx_E9_5_rep1__24\"\n",
      "[18] \"pharynx_atlas__CR_pharynx_E9_5_rep1__25\"\n",
      "[19] \"pharynx_atlas__CR_pharynx_E9_5_rep1__26\"\n",
      "[20] \"pharynx_atlas__CR_pharynx_E9_5_rep1__27\"\n",
      "[21] \"pharynx_atlas__CR_pharynx_E9_5_rep1__28\"\n",
      "[22] \"pharynx_atlas__CR_pharynx_E9_5_rep1__29\"\n",
      "[23] \"pharynx_atlas__CR_pharynx_E9_5_rep1__3\" \n",
      "[24] \"pharynx_atlas__CR_pharynx_E9_5_rep1__30\"\n",
      "[25] \"pharynx_atlas__CR_pharynx_E9_5_rep1__31\"\n",
      "[26] \"pharynx_atlas__CR_pharynx_E9_5_rep1__32\"\n",
      "[27] \"pharynx_atlas__CR_pharynx_E9_5_rep1__33\"\n",
      "[28] \"pharynx_atlas__CR_pharynx_E9_5_rep1__34\"\n",
      "[29] \"pharynx_atlas__CR_pharynx_E9_5_rep1__35\"\n",
      "[30] \"pharynx_atlas__CR_pharynx_E9_5_rep1__36\"\n",
      "[31] \"pharynx_atlas__CR_pharynx_E9_5_rep1__37\"\n",
      "[32] \"pharynx_atlas__CR_pharynx_E9_5_rep1__38\"\n",
      "[33] \"pharynx_atlas__CR_pharynx_E9_5_rep1__39\"\n",
      "[34] \"pharynx_atlas__CR_pharynx_E9_5_rep1__4\" \n",
      "[35] \"pharynx_atlas__CR_pharynx_E9_5_rep1__40\"\n",
      "[36] \"pharynx_atlas__CR_pharynx_E9_5_rep1__41\"\n",
      "[37] \"pharynx_atlas__CR_pharynx_E9_5_rep1__42\"\n",
      "[38] \"pharynx_atlas__CR_pharynx_E9_5_rep1__43\"\n",
      "[39] \"pharynx_atlas__CR_pharynx_E9_5_rep1__44\"\n",
      "[40] \"pharynx_atlas__CR_pharynx_E9_5_rep1__45\"\n",
      "[41] \"pharynx_atlas__CR_pharynx_E9_5_rep1__46\"\n",
      "[42] \"pharynx_atlas__CR_pharynx_E9_5_rep1__47\"\n",
      "[43] \"pharynx_atlas__CR_pharynx_E9_5_rep1__48\"\n",
      "[44] \"pharynx_atlas__CR_pharynx_E9_5_rep1__49\"\n",
      "[45] \"pharynx_atlas__CR_pharynx_E9_5_rep1__5\" \n",
      "[46] \"pharynx_atlas__CR_pharynx_E9_5_rep1__50\"\n",
      "[47] \"pharynx_atlas__CR_pharynx_E9_5_rep1__51\"\n",
      "[48] \"pharynx_atlas__CR_pharynx_E9_5_rep1__52\"\n",
      "[49] \"pharynx_atlas__CR_pharynx_E9_5_rep1__53\"\n",
      "[50] \"pharynx_atlas__CR_pharynx_E9_5_rep1__6\" \n",
      "[51] \"pharynx_atlas__CR_pharynx_E9_5_rep1__7\" \n",
      "[52] \"pharynx_atlas__CR_pharynx_E9_5_rep1__8\" \n",
      "[53] \"pharynx_atlas__CR_pharynx_E9_5_rep1__9\" \n",
      "[1] 27998    53\n",
      "[1] 27998    53\n",
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/data/pharynx_atlas/metacell/CR_pharynx_E9_5_rep2\"\n",
      " [1] \"pharynx_atlas__CR_pharynx_E9_5_rep2__1\" \n",
      " [2] \"pharynx_atlas__CR_pharynx_E9_5_rep2__10\"\n",
      " [3] \"pharynx_atlas__CR_pharynx_E9_5_rep2__11\"\n",
      " [4] \"pharynx_atlas__CR_pharynx_E9_5_rep2__12\"\n",
      " [5] \"pharynx_atlas__CR_pharynx_E9_5_rep2__13\"\n",
      " [6] \"pharynx_atlas__CR_pharynx_E9_5_rep2__14\"\n",
      " [7] \"pharynx_atlas__CR_pharynx_E9_5_rep2__15\"\n",
      " [8] \"pharynx_atlas__CR_pharynx_E9_5_rep2__16\"\n",
      " [9] \"pharynx_atlas__CR_pharynx_E9_5_rep2__17\"\n",
      "[10] \"pharynx_atlas__CR_pharynx_E9_5_rep2__18\"\n",
      "[11] \"pharynx_atlas__CR_pharynx_E9_5_rep2__19\"\n",
      "[12] \"pharynx_atlas__CR_pharynx_E9_5_rep2__2\" \n",
      "[13] \"pharynx_atlas__CR_pharynx_E9_5_rep2__20\"\n",
      "[14] \"pharynx_atlas__CR_pharynx_E9_5_rep2__21\"\n",
      "[15] \"pharynx_atlas__CR_pharynx_E9_5_rep2__22\"\n",
      "[16] \"pharynx_atlas__CR_pharynx_E9_5_rep2__23\"\n",
      "[17] \"pharynx_atlas__CR_pharynx_E9_5_rep2__24\"\n",
      "[18] \"pharynx_atlas__CR_pharynx_E9_5_rep2__25\"\n",
      "[19] \"pharynx_atlas__CR_pharynx_E9_5_rep2__26\"\n",
      "[20] \"pharynx_atlas__CR_pharynx_E9_5_rep2__27\"\n",
      "[21] \"pharynx_atlas__CR_pharynx_E9_5_rep2__28\"\n",
      "[22] \"pharynx_atlas__CR_pharynx_E9_5_rep2__29\"\n",
      "[23] \"pharynx_atlas__CR_pharynx_E9_5_rep2__3\" \n",
      "[24] \"pharynx_atlas__CR_pharynx_E9_5_rep2__30\"\n",
      "[25] \"pharynx_atlas__CR_pharynx_E9_5_rep2__31\"\n",
      "[26] \"pharynx_atlas__CR_pharynx_E9_5_rep2__32\"\n",
      "[27] \"pharynx_atlas__CR_pharynx_E9_5_rep2__33\"\n",
      "[28] \"pharynx_atlas__CR_pharynx_E9_5_rep2__34\"\n",
      "[29] \"pharynx_atlas__CR_pharynx_E9_5_rep2__35\"\n",
      "[30] \"pharynx_atlas__CR_pharynx_E9_5_rep2__36\"\n",
      "[31] \"pharynx_atlas__CR_pharynx_E9_5_rep2__37\"\n",
      "[32] \"pharynx_atlas__CR_pharynx_E9_5_rep2__38\"\n",
      "[33] \"pharynx_atlas__CR_pharynx_E9_5_rep2__39\"\n",
      "[34] \"pharynx_atlas__CR_pharynx_E9_5_rep2__4\" \n",
      "[35] \"pharynx_atlas__CR_pharynx_E9_5_rep2__40\"\n",
      "[36] \"pharynx_atlas__CR_pharynx_E9_5_rep2__41\"\n",
      "[37] \"pharynx_atlas__CR_pharynx_E9_5_rep2__42\"\n",
      "[38] \"pharynx_atlas__CR_pharynx_E9_5_rep2__43\"\n",
      "[39] \"pharynx_atlas__CR_pharynx_E9_5_rep2__44\"\n",
      "[40] \"pharynx_atlas__CR_pharynx_E9_5_rep2__45\"\n",
      "[41] \"pharynx_atlas__CR_pharynx_E9_5_rep2__46\"\n",
      "[42] \"pharynx_atlas__CR_pharynx_E9_5_rep2__47\"\n",
      "[43] \"pharynx_atlas__CR_pharynx_E9_5_rep2__48\"\n",
      "[44] \"pharynx_atlas__CR_pharynx_E9_5_rep2__49\"\n",
      "[45] \"pharynx_atlas__CR_pharynx_E9_5_rep2__5\" \n",
      "[46] \"pharynx_atlas__CR_pharynx_E9_5_rep2__50\"\n",
      "[47] \"pharynx_atlas__CR_pharynx_E9_5_rep2__51\"\n",
      "[48] \"pharynx_atlas__CR_pharynx_E9_5_rep2__52\"\n",
      "[49] \"pharynx_atlas__CR_pharynx_E9_5_rep2__53\"\n",
      "[50] \"pharynx_atlas__CR_pharynx_E9_5_rep2__54\"\n",
      "[51] \"pharynx_atlas__CR_pharynx_E9_5_rep2__55\"\n",
      "[52] \"pharynx_atlas__CR_pharynx_E9_5_rep2__56\"\n",
      "[53] \"pharynx_atlas__CR_pharynx_E9_5_rep2__57\"\n",
      "[54] \"pharynx_atlas__CR_pharynx_E9_5_rep2__58\"\n",
      "[55] \"pharynx_atlas__CR_pharynx_E9_5_rep2__59\"\n",
      "[56] \"pharynx_atlas__CR_pharynx_E9_5_rep2__6\" \n",
      "[57] \"pharynx_atlas__CR_pharynx_E9_5_rep2__60\"\n",
      "[58] \"pharynx_atlas__CR_pharynx_E9_5_rep2__61\"\n",
      "[59] \"pharynx_atlas__CR_pharynx_E9_5_rep2__62\"\n",
      "[60] \"pharynx_atlas__CR_pharynx_E9_5_rep2__63\"\n",
      "[61] \"pharynx_atlas__CR_pharynx_E9_5_rep2__64\"\n",
      "[62] \"pharynx_atlas__CR_pharynx_E9_5_rep2__65\"\n",
      "[63] \"pharynx_atlas__CR_pharynx_E9_5_rep2__66\"\n",
      "[64] \"pharynx_atlas__CR_pharynx_E9_5_rep2__67\"\n",
      "[65] \"pharynx_atlas__CR_pharynx_E9_5_rep2__68\"\n",
      "[66] \"pharynx_atlas__CR_pharynx_E9_5_rep2__69\"\n",
      "[67] \"pharynx_atlas__CR_pharynx_E9_5_rep2__7\" \n",
      "[68] \"pharynx_atlas__CR_pharynx_E9_5_rep2__70\"\n",
      "[69] \"pharynx_atlas__CR_pharynx_E9_5_rep2__71\"\n",
      "[70] \"pharynx_atlas__CR_pharynx_E9_5_rep2__72\"\n",
      "[71] \"pharynx_atlas__CR_pharynx_E9_5_rep2__73\"\n",
      "[72] \"pharynx_atlas__CR_pharynx_E9_5_rep2__74\"\n",
      "[73] \"pharynx_atlas__CR_pharynx_E9_5_rep2__8\" \n",
      "[74] \"pharynx_atlas__CR_pharynx_E9_5_rep2__9\" \n",
      "[1] 27998    74\n",
      "[1] 27998    74\n",
      "[1] \"Finished merging dge matrix CR_pharynx_E10_5_rep5\"\n",
      "[1] \"Finished merging dge matrix CR_pharynx_E10_5_rep6\"\n",
      "[1] \"Finished merging dge matrix CR_pharynx_E10_5_rep7\"\n",
      "[1] \"Finished merging dge matrix CR_pharynx_E11_5_rep2\"\n",
      "[1] \"Finished merging dge matrix CR_pharynx_E11_5_rep2B\"\n",
      "[1] \"Finished merging dge matrix CR_pharynx_E11_5_rep3\"\n",
      "[1] \"Finished merging dge matrix CR_pharynx_E12_5_rep1\"\n",
      "[1] \"Finished merging dge matrix CR_pharynx_E12_5_rep2\"\n",
      "[1] \"Finished merging dge matrix CR_pharynx_E9_5_rep1\"\n",
      "[1] \"Finished merging dge matrix CR_pharynx_E9_5_rep2\"\n"
     ]
    }
   ],
   "source": [
    "metacells = iterate_within_source( dataset_name = \"pharynx_atlas\",\n",
    "                                       FUN = load_expr_by_dge,\n",
    "                                       mode = \"simple\",\n",
    "                                       type = \"metacell\",\n",
    "                                       test = F ) %>% dge_merge_list() %>% as.matrix\n",
    "  \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e6a4b405",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(rownames(metacells),\"tmp1_1.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "ca811903",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Maximum value in the expression matrix: 6538\n",
      "\n",
      "Ratio of detected vs non-detected: 0.52\n",
      "\n",
      "Number of counts (in the dataset units) per gene:\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
      "      0       1      66    2187    1222  557230 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Number of cells in which each gene is detected:\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
      "    0.0     1.0    55.0   188.1   422.0   551.0 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of genes left after applying the following filters (sequential):\n",
      "\n",
      "\t17584\tgenes with counts per gene > 10\n",
      "\n",
      "\t17584\tgenes detected in more than 2 cells\n",
      "\n",
      "\t15329\tgenes available in RcisTarget database\n",
      "\n",
      "Gene list saved in int/1.1_genesKept.Rds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "genes_keep <- geneFiltering(metacells, \n",
    "                              scenicOptions=scenicOptions,\n",
    "                              minCountsPerGene=10,\n",
    "                              minSamples=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5b4f3078",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(genes_keep,\"tmp2_1.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "b7373b21",
   "metadata": {},
   "outputs": [],
   "source": [
    "metacells = metacells[genes_keep, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "901810c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using 1269 TFs as potential regulators...\n",
      "\n",
      "Running GENIE3 part 1\n",
      "\n",
      "Running GENIE3 part 10\n",
      "\n",
      "Running GENIE3 part 2\n",
      "\n",
      "Running GENIE3 part 3\n",
      "\n",
      "Running GENIE3 part 4\n",
      "\n",
      "Running GENIE3 part 5\n",
      "\n",
      "Running GENIE3 part 6\n",
      "\n",
      "Running GENIE3 part 7\n",
      "\n",
      "Running GENIE3 part 8\n",
      "\n",
      "Running GENIE3 part 9\n",
      "\n",
      "Finished running GENIE3.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "tf_target = runGenie3(metacells, scenicOptions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "8ce72d9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3672138 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>TF</th><th scope=col>Target</th><th scope=col>weight</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Hoxb6 </td><td>Hoxb5os      </td><td>0.10228378</td></tr>\n",
       "\t<tr><td>Tfap2a</td><td>Gm26688      </td><td>0.10142356</td></tr>\n",
       "\t<tr><td>Esrrg </td><td>Gcm2         </td><td>0.09993458</td></tr>\n",
       "\t<tr><td>Tfap2a</td><td>Tnfrsf11b    </td><td>0.09987369</td></tr>\n",
       "\t<tr><td>Tfap2b</td><td>Cxcl14       </td><td>0.09715990</td></tr>\n",
       "\t<tr><td>Tfap2a</td><td>Wnt9b        </td><td>0.09483255</td></tr>\n",
       "\t<tr><td>Celf4 </td><td>Anxa1        </td><td>0.09125103</td></tr>\n",
       "\t<tr><td>Sim1  </td><td>Otx2os1      </td><td>0.08997949</td></tr>\n",
       "\t<tr><td>Esrrg </td><td>Stac2        </td><td>0.08891674</td></tr>\n",
       "\t<tr><td>Pax8  </td><td>Kcnj16       </td><td>0.08712814</td></tr>\n",
       "\t<tr><td>Gli3  </td><td>Rarres1      </td><td>0.08622034</td></tr>\n",
       "\t<tr><td>Rfx6  </td><td>Hnf1b        </td><td>0.08610888</td></tr>\n",
       "\t<tr><td>Tfap2b</td><td>Wnt3         </td><td>0.08407824</td></tr>\n",
       "\t<tr><td>Otx2  </td><td>Otx2os1      </td><td>0.08343711</td></tr>\n",
       "\t<tr><td>Otx2  </td><td>Postn        </td><td>0.08314272</td></tr>\n",
       "\t<tr><td>Tfap2a</td><td>Tfap2b       </td><td>0.08197262</td></tr>\n",
       "\t<tr><td>Tfap2a</td><td>Hgf          </td><td>0.08183546</td></tr>\n",
       "\t<tr><td>Esrrg </td><td>Casr         </td><td>0.07952289</td></tr>\n",
       "\t<tr><td>Rfx6  </td><td>Enpp1        </td><td>0.07945111</td></tr>\n",
       "\t<tr><td>Terf1 </td><td>Cdkn3        </td><td>0.07899770</td></tr>\n",
       "\t<tr><td>Fli1  </td><td>Pde1a        </td><td>0.07897190</td></tr>\n",
       "\t<tr><td>Esrrg </td><td>Fam20a       </td><td>0.07849946</td></tr>\n",
       "\t<tr><td>Terf1 </td><td>Cenpf        </td><td>0.07813805</td></tr>\n",
       "\t<tr><td>Gcm2  </td><td>Stac2        </td><td>0.07795471</td></tr>\n",
       "\t<tr><td>Terf1 </td><td>Ccnb1        </td><td>0.07760936</td></tr>\n",
       "\t<tr><td>Rfx6  </td><td>Gata6        </td><td>0.07690102</td></tr>\n",
       "\t<tr><td>Rfx6  </td><td>2410017I17Rik</td><td>0.07682477</td></tr>\n",
       "\t<tr><td>Tfap2a</td><td>Wnt3         </td><td>0.07676816</td></tr>\n",
       "\t<tr><td>Rfx6  </td><td>Slc17a2      </td><td>0.07668695</td></tr>\n",
       "\t<tr><td>Terf1 </td><td>Arl6ip1      </td><td>0.07574987</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>Apex2  </td><td>Ushbp1        </td><td>0.001000006</td></tr>\n",
       "\t<tr><td>Arid3b </td><td>Eif2s3x       </td><td>0.001000006</td></tr>\n",
       "\t<tr><td>Zfp131 </td><td>Rdm1          </td><td>0.001000006</td></tr>\n",
       "\t<tr><td>Npas2  </td><td>Klf7          </td><td>0.001000006</td></tr>\n",
       "\t<tr><td>Smap2  </td><td>Prrx1         </td><td>0.001000005</td></tr>\n",
       "\t<tr><td>Zbtb7b </td><td>Mapre3        </td><td>0.001000005</td></tr>\n",
       "\t<tr><td>Prox1  </td><td>Phb2          </td><td>0.001000005</td></tr>\n",
       "\t<tr><td>Smad6  </td><td>Hkdc1         </td><td>0.001000004</td></tr>\n",
       "\t<tr><td>Klf6   </td><td>Bcl2l14       </td><td>0.001000004</td></tr>\n",
       "\t<tr><td>Zfp384 </td><td>Nxpe5         </td><td>0.001000004</td></tr>\n",
       "\t<tr><td>Zfp110 </td><td>2610044O15Rik8</td><td>0.001000004</td></tr>\n",
       "\t<tr><td>Zfp62  </td><td>C130083M11Rik </td><td>0.001000004</td></tr>\n",
       "\t<tr><td>Mef2c  </td><td>Capns1        </td><td>0.001000004</td></tr>\n",
       "\t<tr><td>Pick1  </td><td>Pde2a         </td><td>0.001000004</td></tr>\n",
       "\t<tr><td>Zfp362 </td><td>Oc90          </td><td>0.001000003</td></tr>\n",
       "\t<tr><td>Hnrnph3</td><td>Hdac11        </td><td>0.001000003</td></tr>\n",
       "\t<tr><td>Sp4    </td><td>Ttc1          </td><td>0.001000003</td></tr>\n",
       "\t<tr><td>Runx2  </td><td>Klhdc8a       </td><td>0.001000003</td></tr>\n",
       "\t<tr><td>Setbp1 </td><td>Dcaf13        </td><td>0.001000003</td></tr>\n",
       "\t<tr><td>Ppara  </td><td>Zfp600        </td><td>0.001000002</td></tr>\n",
       "\t<tr><td>Mga    </td><td>Gm1976        </td><td>0.001000002</td></tr>\n",
       "\t<tr><td>Zbtb11 </td><td>Them4         </td><td>0.001000002</td></tr>\n",
       "\t<tr><td>E2f5   </td><td>Ackr4         </td><td>0.001000002</td></tr>\n",
       "\t<tr><td>Lrrfip1</td><td>Col6a2        </td><td>0.001000002</td></tr>\n",
       "\t<tr><td>Tead1  </td><td>Fam183b       </td><td>0.001000001</td></tr>\n",
       "\t<tr><td>Snd1   </td><td>Pik3r2        </td><td>0.001000001</td></tr>\n",
       "\t<tr><td>Rcor1  </td><td>Rfxap         </td><td>0.001000001</td></tr>\n",
       "\t<tr><td>Dab2   </td><td>2310043O21Rik </td><td>0.001000001</td></tr>\n",
       "\t<tr><td>Nkx2-1 </td><td>Arhgap20os    </td><td>0.001000001</td></tr>\n",
       "\t<tr><td>Zfp654 </td><td>AU019990      </td><td>0.001000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3672138 × 3\n",
       "\\begin{tabular}{lll}\n",
       " TF & Target & weight\\\\\n",
       " <fct> & <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Hoxb6  & Hoxb5os       & 0.10228378\\\\\n",
       "\t Tfap2a & Gm26688       & 0.10142356\\\\\n",
       "\t Esrrg  & Gcm2          & 0.09993458\\\\\n",
       "\t Tfap2a & Tnfrsf11b     & 0.09987369\\\\\n",
       "\t Tfap2b & Cxcl14        & 0.09715990\\\\\n",
       "\t Tfap2a & Wnt9b         & 0.09483255\\\\\n",
       "\t Celf4  & Anxa1         & 0.09125103\\\\\n",
       "\t Sim1   & Otx2os1       & 0.08997949\\\\\n",
       "\t Esrrg  & Stac2         & 0.08891674\\\\\n",
       "\t Pax8   & Kcnj16        & 0.08712814\\\\\n",
       "\t Gli3   & Rarres1       & 0.08622034\\\\\n",
       "\t Rfx6   & Hnf1b         & 0.08610888\\\\\n",
       "\t Tfap2b & Wnt3          & 0.08407824\\\\\n",
       "\t Otx2   & Otx2os1       & 0.08343711\\\\\n",
       "\t Otx2   & Postn         & 0.08314272\\\\\n",
       "\t Tfap2a & Tfap2b        & 0.08197262\\\\\n",
       "\t Tfap2a & Hgf           & 0.08183546\\\\\n",
       "\t Esrrg  & Casr          & 0.07952289\\\\\n",
       "\t Rfx6   & Enpp1         & 0.07945111\\\\\n",
       "\t Terf1  & Cdkn3         & 0.07899770\\\\\n",
       "\t Fli1   & Pde1a         & 0.07897190\\\\\n",
       "\t Esrrg  & Fam20a        & 0.07849946\\\\\n",
       "\t Terf1  & Cenpf         & 0.07813805\\\\\n",
       "\t Gcm2   & Stac2         & 0.07795471\\\\\n",
       "\t Terf1  & Ccnb1         & 0.07760936\\\\\n",
       "\t Rfx6   & Gata6         & 0.07690102\\\\\n",
       "\t Rfx6   & 2410017I17Rik & 0.07682477\\\\\n",
       "\t Tfap2a & Wnt3          & 0.07676816\\\\\n",
       "\t Rfx6   & Slc17a2       & 0.07668695\\\\\n",
       "\t Terf1  & Arl6ip1       & 0.07574987\\\\\n",
       "\t ⋮ & ⋮ & ⋮\\\\\n",
       "\t Apex2   & Ushbp1         & 0.001000006\\\\\n",
       "\t Arid3b  & Eif2s3x        & 0.001000006\\\\\n",
       "\t Zfp131  & Rdm1           & 0.001000006\\\\\n",
       "\t Npas2   & Klf7           & 0.001000006\\\\\n",
       "\t Smap2   & Prrx1          & 0.001000005\\\\\n",
       "\t Zbtb7b  & Mapre3         & 0.001000005\\\\\n",
       "\t Prox1   & Phb2           & 0.001000005\\\\\n",
       "\t Smad6   & Hkdc1          & 0.001000004\\\\\n",
       "\t Klf6    & Bcl2l14        & 0.001000004\\\\\n",
       "\t Zfp384  & Nxpe5          & 0.001000004\\\\\n",
       "\t Zfp110  & 2610044O15Rik8 & 0.001000004\\\\\n",
       "\t Zfp62   & C130083M11Rik  & 0.001000004\\\\\n",
       "\t Mef2c   & Capns1         & 0.001000004\\\\\n",
       "\t Pick1   & Pde2a          & 0.001000004\\\\\n",
       "\t Zfp362  & Oc90           & 0.001000003\\\\\n",
       "\t Hnrnph3 & Hdac11         & 0.001000003\\\\\n",
       "\t Sp4     & Ttc1           & 0.001000003\\\\\n",
       "\t Runx2   & Klhdc8a        & 0.001000003\\\\\n",
       "\t Setbp1  & Dcaf13         & 0.001000003\\\\\n",
       "\t Ppara   & Zfp600         & 0.001000002\\\\\n",
       "\t Mga     & Gm1976         & 0.001000002\\\\\n",
       "\t Zbtb11  & Them4          & 0.001000002\\\\\n",
       "\t E2f5    & Ackr4          & 0.001000002\\\\\n",
       "\t Lrrfip1 & Col6a2         & 0.001000002\\\\\n",
       "\t Tead1   & Fam183b        & 0.001000001\\\\\n",
       "\t Snd1    & Pik3r2         & 0.001000001\\\\\n",
       "\t Rcor1   & Rfxap          & 0.001000001\\\\\n",
       "\t Dab2    & 2310043O21Rik  & 0.001000001\\\\\n",
       "\t Nkx2-1  & Arhgap20os     & 0.001000001\\\\\n",
       "\t Zfp654  & AU019990       & 0.001000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3672138 × 3\n",
       "\n",
       "| TF &lt;fct&gt; | Target &lt;fct&gt; | weight &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| Hoxb6  | Hoxb5os       | 0.10228378 |\n",
       "| Tfap2a | Gm26688       | 0.10142356 |\n",
       "| Esrrg  | Gcm2          | 0.09993458 |\n",
       "| Tfap2a | Tnfrsf11b     | 0.09987369 |\n",
       "| Tfap2b | Cxcl14        | 0.09715990 |\n",
       "| Tfap2a | Wnt9b         | 0.09483255 |\n",
       "| Celf4  | Anxa1         | 0.09125103 |\n",
       "| Sim1   | Otx2os1       | 0.08997949 |\n",
       "| Esrrg  | Stac2         | 0.08891674 |\n",
       "| Pax8   | Kcnj16        | 0.08712814 |\n",
       "| Gli3   | Rarres1       | 0.08622034 |\n",
       "| Rfx6   | Hnf1b         | 0.08610888 |\n",
       "| Tfap2b | Wnt3          | 0.08407824 |\n",
       "| Otx2   | Otx2os1       | 0.08343711 |\n",
       "| Otx2   | Postn         | 0.08314272 |\n",
       "| Tfap2a | Tfap2b        | 0.08197262 |\n",
       "| Tfap2a | Hgf           | 0.08183546 |\n",
       "| Esrrg  | Casr          | 0.07952289 |\n",
       "| Rfx6   | Enpp1         | 0.07945111 |\n",
       "| Terf1  | Cdkn3         | 0.07899770 |\n",
       "| Fli1   | Pde1a         | 0.07897190 |\n",
       "| Esrrg  | Fam20a        | 0.07849946 |\n",
       "| Terf1  | Cenpf         | 0.07813805 |\n",
       "| Gcm2   | Stac2         | 0.07795471 |\n",
       "| Terf1  | Ccnb1         | 0.07760936 |\n",
       "| Rfx6   | Gata6         | 0.07690102 |\n",
       "| Rfx6   | 2410017I17Rik | 0.07682477 |\n",
       "| Tfap2a | Wnt3          | 0.07676816 |\n",
       "| Rfx6   | Slc17a2       | 0.07668695 |\n",
       "| Terf1  | Arl6ip1       | 0.07574987 |\n",
       "| ⋮ | ⋮ | ⋮ |\n",
       "| Apex2   | Ushbp1         | 0.001000006 |\n",
       "| Arid3b  | Eif2s3x        | 0.001000006 |\n",
       "| Zfp131  | Rdm1           | 0.001000006 |\n",
       "| Npas2   | Klf7           | 0.001000006 |\n",
       "| Smap2   | Prrx1          | 0.001000005 |\n",
       "| Zbtb7b  | Mapre3         | 0.001000005 |\n",
       "| Prox1   | Phb2           | 0.001000005 |\n",
       "| Smad6   | Hkdc1          | 0.001000004 |\n",
       "| Klf6    | Bcl2l14        | 0.001000004 |\n",
       "| Zfp384  | Nxpe5          | 0.001000004 |\n",
       "| Zfp110  | 2610044O15Rik8 | 0.001000004 |\n",
       "| Zfp62   | C130083M11Rik  | 0.001000004 |\n",
       "| Mef2c   | Capns1         | 0.001000004 |\n",
       "| Pick1   | Pde2a          | 0.001000004 |\n",
       "| Zfp362  | Oc90           | 0.001000003 |\n",
       "| Hnrnph3 | Hdac11         | 0.001000003 |\n",
       "| Sp4     | Ttc1           | 0.001000003 |\n",
       "| Runx2   | Klhdc8a        | 0.001000003 |\n",
       "| Setbp1  | Dcaf13         | 0.001000003 |\n",
       "| Ppara   | Zfp600         | 0.001000002 |\n",
       "| Mga     | Gm1976         | 0.001000002 |\n",
       "| Zbtb11  | Them4          | 0.001000002 |\n",
       "| E2f5    | Ackr4          | 0.001000002 |\n",
       "| Lrrfip1 | Col6a2         | 0.001000002 |\n",
       "| Tead1   | Fam183b        | 0.001000001 |\n",
       "| Snd1    | Pik3r2         | 0.001000001 |\n",
       "| Rcor1   | Rfxap          | 0.001000001 |\n",
       "| Dab2    | 2310043O21Rik  | 0.001000001 |\n",
       "| Nkx2-1  | Arhgap20os     | 0.001000001 |\n",
       "| Zfp654  | AU019990       | 0.001000000 |\n",
       "\n"
      ],
      "text/plain": [
       "        TF      Target         weight     \n",
       "1       Hoxb6   Hoxb5os        0.10228378 \n",
       "2       Tfap2a  Gm26688        0.10142356 \n",
       "3       Esrrg   Gcm2           0.09993458 \n",
       "4       Tfap2a  Tnfrsf11b      0.09987369 \n",
       "5       Tfap2b  Cxcl14         0.09715990 \n",
       "6       Tfap2a  Wnt9b          0.09483255 \n",
       "7       Celf4   Anxa1          0.09125103 \n",
       "8       Sim1    Otx2os1        0.08997949 \n",
       "9       Esrrg   Stac2          0.08891674 \n",
       "10      Pax8    Kcnj16         0.08712814 \n",
       "11      Gli3    Rarres1        0.08622034 \n",
       "12      Rfx6    Hnf1b          0.08610888 \n",
       "13      Tfap2b  Wnt3           0.08407824 \n",
       "14      Otx2    Otx2os1        0.08343711 \n",
       "15      Otx2    Postn          0.08314272 \n",
       "16      Tfap2a  Tfap2b         0.08197262 \n",
       "17      Tfap2a  Hgf            0.08183546 \n",
       "18      Esrrg   Casr           0.07952289 \n",
       "19      Rfx6    Enpp1          0.07945111 \n",
       "20      Terf1   Cdkn3          0.07899770 \n",
       "21      Fli1    Pde1a          0.07897190 \n",
       "22      Esrrg   Fam20a         0.07849946 \n",
       "23      Terf1   Cenpf          0.07813805 \n",
       "24      Gcm2    Stac2          0.07795471 \n",
       "25      Terf1   Ccnb1          0.07760936 \n",
       "26      Rfx6    Gata6          0.07690102 \n",
       "27      Rfx6    2410017I17Rik  0.07682477 \n",
       "28      Tfap2a  Wnt3           0.07676816 \n",
       "29      Rfx6    Slc17a2        0.07668695 \n",
       "30      Terf1   Arl6ip1        0.07574987 \n",
       "⋮       ⋮       ⋮              ⋮          \n",
       "3672109 Apex2   Ushbp1         0.001000006\n",
       "3672110 Arid3b  Eif2s3x        0.001000006\n",
       "3672111 Zfp131  Rdm1           0.001000006\n",
       "3672112 Npas2   Klf7           0.001000006\n",
       "3672113 Smap2   Prrx1          0.001000005\n",
       "3672114 Zbtb7b  Mapre3         0.001000005\n",
       "3672115 Prox1   Phb2           0.001000005\n",
       "3672116 Smad6   Hkdc1          0.001000004\n",
       "3672117 Klf6    Bcl2l14        0.001000004\n",
       "3672118 Zfp384  Nxpe5          0.001000004\n",
       "3672119 Zfp110  2610044O15Rik8 0.001000004\n",
       "3672120 Zfp62   C130083M11Rik  0.001000004\n",
       "3672121 Mef2c   Capns1         0.001000004\n",
       "3672122 Pick1   Pde2a          0.001000004\n",
       "3672123 Zfp362  Oc90           0.001000003\n",
       "3672124 Hnrnph3 Hdac11         0.001000003\n",
       "3672125 Sp4     Ttc1           0.001000003\n",
       "3672126 Runx2   Klhdc8a        0.001000003\n",
       "3672127 Setbp1  Dcaf13         0.001000003\n",
       "3672128 Ppara   Zfp600         0.001000002\n",
       "3672129 Mga     Gm1976         0.001000002\n",
       "3672130 Zbtb11  Them4          0.001000002\n",
       "3672131 E2f5    Ackr4          0.001000002\n",
       "3672132 Lrrfip1 Col6a2         0.001000002\n",
       "3672133 Tead1   Fam183b        0.001000001\n",
       "3672134 Snd1    Pik3r2         0.001000001\n",
       "3672135 Rcor1   Rfxap          0.001000001\n",
       "3672136 Dab2    2310043O21Rik  0.001000001\n",
       "3672137 Nkx2-1  Arhgap20os     0.001000001\n",
       "3672138 Zfp654  AU019990       0.001000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tf_target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "1489c6ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(tf_target['Target'],\"tmp3_2.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "14f07de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(tf_target['TF'],\"tmp3_1.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e7c55da2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Load source metadata\n",
    "#' \n",
    "#' \n",
    "#' Returns a dataframe\n",
    "load_grn_metadata = function( complete_only = T ){\n",
    "  metadata_df = read.csv(\"published_networks.csv\", stringsAsFactors = F)\n",
    "  rownames(metadata_df) = metadata_df[[\"name\"]]\n",
    "  if(complete_only){ metadata_df = subset(metadata_df, is_ready == \"yes\")}\n",
    "  metadata_df\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "3821e504",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Save a new gene regulatory (sub-)network.\n",
    "#' \n",
    "#' @param grn_name @param dge_name This function saves to here(\"networks\", grn_name, \"networks\", subnetwork_name).\n",
    "#' \n",
    "save_grn_by_subnetwork = function( grn_name, subnetwork_name, X, ... ){\n",
    "  grn_location = here(\"networks\", grn_name, \"networks\", subnetwork_name)\n",
    "  cat(\"Attempting to print relevant metadata. You may need to edit metadata by hand. \\n\")\n",
    "  load_grn_metadata() %>% subset(name == grn_name) %>% print\n",
    "  write.table(X,\n",
    "              grn_location,\n",
    "              sep = \",\", col.names = F, row.names = F)\n",
    "  cat(\"Wrote grn to \", grn_location, \"\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "b2971822",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to print relevant metadata. You may need to edit metadata by hand. \n",
      "                         name\n",
      "scenic_pharynx scenic_pharynx\n",
      "                                                       description\n",
      "scenic_pharynx SCENIC applied on metacells from the pharynx atlas \n",
      "                                  data_used species         link     citation\n",
      "scenic_pharynx pharynx atlas data freeze 13   mouse This project This project\n",
      "                                                       version is_ready\n",
      "scenic_pharynx Commit 1e116eb48c157a52cb447bbfd9e38eb7ef17242b      yes\n",
      "               first_column\n",
      "scenic_pharynx    regulator\n",
      "Wrote grn to  /Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/networks/scenic_pharynx/networks/pharynx_atlas \n"
     ]
    }
   ],
   "source": [
    "save_grn_by_subnetwork(\"scenic_pharynx\", \"pharynx_atlas\", tf_target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "6799267a",
   "metadata": {},
   "outputs": [],
   "source": [
    "scenic_filtered = subset(tf_target, weight >= 0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "28fa7960",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attempting to print relevant metadata. You may need to edit metadata by hand. \n",
      "                         name\n",
      "scenic_pharynx scenic_pharynx\n",
      "                                                       description\n",
      "scenic_pharynx SCENIC applied on metacells from the pharynx atlas \n",
      "                                  data_used species         link     citation\n",
      "scenic_pharynx pharynx atlas data freeze 13   mouse This project This project\n",
      "                                                       version is_ready\n",
      "scenic_pharynx Commit 1e116eb48c157a52cb447bbfd9e38eb7ef17242b      yes\n",
      "               first_column\n",
      "scenic_pharynx    regulator\n",
      "Wrote grn to  /Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/networks/scenic_pharynx/networks/pharynx_atlas_filtered \n"
     ]
    }
   ],
   "source": [
    "save_grn_by_subnetwork(\"scenic_pharynx\", paste0(\"pharynx_atlas\", \"_filtered\"), scenic_filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "f398d4ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "EXPECTED_GRN_COLNAMES = c(\"regulator\", \"target\", \"weight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "4d6bd23d",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 116623 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>TF</th><th scope=col>Target</th><th scope=col>weight</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Hoxb6 </td><td>Hoxb5os      </td><td>0.11868356</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>Tfap2a</td><td>Gm26688      </td><td>0.11001707</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>Rfx6  </td><td>Hnf1b        </td><td>0.09558139</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Sim1  </td><td>Otx2os1      </td><td>0.09500712</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>Tfap2a</td><td>Tnfrsf11b    </td><td>0.09362741</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>Celf4 </td><td>Anxa1        </td><td>0.09282246</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>Tfap2a</td><td>Tfap2b       </td><td>0.09205630</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>Terf1 </td><td>Pif1         </td><td>0.09076750</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>Esrrg </td><td>Stac2        </td><td>0.09070625</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>Tfap2b</td><td>Wnt3         </td><td>0.09022312</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>Terf1 </td><td>Lockd        </td><td>0.08919815</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>Terf1 </td><td>Cdkn3        </td><td>0.08498176</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>Otx2  </td><td>Otx2os1      </td><td>0.08487994</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>Rfx6  </td><td>Enpp1        </td><td>0.08339922</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>Esrrg </td><td>Fam20a       </td><td>0.08308780</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>Sim1  </td><td>Otx2         </td><td>0.08301071</td></tr>\n",
       "\t<tr><th scope=row>17</th><td>Prrx1 </td><td>Kcnk10       </td><td>0.08176160</td></tr>\n",
       "\t<tr><th scope=row>18</th><td>Esrrg </td><td>Casr         </td><td>0.08147499</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>Tfap2b</td><td>Cxcl14       </td><td>0.08043888</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>Tfap2a</td><td>Wnt3         </td><td>0.08020572</td></tr>\n",
       "\t<tr><th scope=row>21</th><td>Esrrg </td><td>Tsc22d3      </td><td>0.07962714</td></tr>\n",
       "\t<tr><th scope=row>22</th><td>Pou6f2</td><td>Vwc2         </td><td>0.07940138</td></tr>\n",
       "\t<tr><th scope=row>23</th><td>Hnf1b </td><td>Rfx6         </td><td>0.07920826</td></tr>\n",
       "\t<tr><th scope=row>24</th><td>Otx2  </td><td>Postn        </td><td>0.07892602</td></tr>\n",
       "\t<tr><th scope=row>25</th><td>Terf1 </td><td>Cenpf        </td><td>0.07885115</td></tr>\n",
       "\t<tr><th scope=row>26</th><td>Terf1 </td><td>Ccnb1        </td><td>0.07878102</td></tr>\n",
       "\t<tr><th scope=row>27</th><td>Tfap2a</td><td>Hgf          </td><td>0.07871124</td></tr>\n",
       "\t<tr><th scope=row>28</th><td>Tfap2b</td><td>Tnfrsf11b    </td><td>0.07800308</td></tr>\n",
       "\t<tr><th scope=row>29</th><td>Tfap2b</td><td>E330013P04Rik</td><td>0.07777844</td></tr>\n",
       "\t<tr><th scope=row>30</th><td>Tfap2b</td><td>Tfap2a       </td><td>0.07746201</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>116594</th><td>Bptf   </td><td>Zbtb44       </td><td>0.01000139</td></tr>\n",
       "\t<tr><th scope=row>116595</th><td>Esrp1  </td><td>Usp20        </td><td>0.01000130</td></tr>\n",
       "\t<tr><th scope=row>116596</th><td>Tcf12  </td><td>Unc13b       </td><td>0.01000119</td></tr>\n",
       "\t<tr><th scope=row>116597</th><td>Thoc2  </td><td>Rcor1        </td><td>0.01000105</td></tr>\n",
       "\t<tr><th scope=row>116598</th><td>Larp1  </td><td>Rcc2         </td><td>0.01000105</td></tr>\n",
       "\t<tr><th scope=row>116599</th><td>Stat5b </td><td>Col22a1      </td><td>0.01000103</td></tr>\n",
       "\t<tr><th scope=row>116600</th><td>Nfatc3 </td><td>Opa1         </td><td>0.01000102</td></tr>\n",
       "\t<tr><th scope=row>116601</th><td>Tfcp2  </td><td>Pou6f1       </td><td>0.01000098</td></tr>\n",
       "\t<tr><th scope=row>116602</th><td>Tcf12  </td><td>Rabep1       </td><td>0.01000098</td></tr>\n",
       "\t<tr><th scope=row>116603</th><td>Fubp1  </td><td>Hn1          </td><td>0.01000096</td></tr>\n",
       "\t<tr><th scope=row>116604</th><td>Fubp1  </td><td>Rcbtb2       </td><td>0.01000093</td></tr>\n",
       "\t<tr><th scope=row>116605</th><td>Cux1   </td><td>Slc37a3      </td><td>0.01000086</td></tr>\n",
       "\t<tr><th scope=row>116606</th><td>Rbms1  </td><td>Pias1        </td><td>0.01000085</td></tr>\n",
       "\t<tr><th scope=row>116607</th><td>Ctbp2  </td><td>Atrn         </td><td>0.01000082</td></tr>\n",
       "\t<tr><th scope=row>116608</th><td>Parp1  </td><td>Glod4        </td><td>0.01000079</td></tr>\n",
       "\t<tr><th scope=row>116609</th><td>Zfp148 </td><td>Ccdc88a      </td><td>0.01000079</td></tr>\n",
       "\t<tr><th scope=row>116610</th><td>Nfatc3 </td><td>Cpne3        </td><td>0.01000077</td></tr>\n",
       "\t<tr><th scope=row>116611</th><td>Zfp652 </td><td>Pkp4         </td><td>0.01000068</td></tr>\n",
       "\t<tr><th scope=row>116612</th><td>Mapk1  </td><td>Rab6a        </td><td>0.01000067</td></tr>\n",
       "\t<tr><th scope=row>116613</th><td>Rbfox2 </td><td>Heatr5b      </td><td>0.01000066</td></tr>\n",
       "\t<tr><th scope=row>116614</th><td>Tbl1xr1</td><td>Oxct1        </td><td>0.01000063</td></tr>\n",
       "\t<tr><th scope=row>116615</th><td>Asap3  </td><td>Snx7         </td><td>0.01000054</td></tr>\n",
       "\t<tr><th scope=row>116616</th><td>Suz12  </td><td>Nudcd2       </td><td>0.01000047</td></tr>\n",
       "\t<tr><th scope=row>116617</th><td>Klf12  </td><td>Lin7a        </td><td>0.01000046</td></tr>\n",
       "\t<tr><th scope=row>116618</th><td>Xpa    </td><td>Cacna1a      </td><td>0.01000030</td></tr>\n",
       "\t<tr><th scope=row>116619</th><td>Rab2a  </td><td>Rai1         </td><td>0.01000027</td></tr>\n",
       "\t<tr><th scope=row>116620</th><td>Zfp521 </td><td>Capsl        </td><td>0.01000020</td></tr>\n",
       "\t<tr><th scope=row>116621</th><td>Lcor   </td><td>Pip4k2c      </td><td>0.01000007</td></tr>\n",
       "\t<tr><th scope=row>116622</th><td>Ikzf2  </td><td>4930480K23Rik</td><td>0.01000004</td></tr>\n",
       "\t<tr><th scope=row>116623</th><td>Nrf1   </td><td>Paip2b       </td><td>0.01000001</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 116623 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & TF & Target & weight\\\\\n",
       "  & <fct> & <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & Hoxb6  & Hoxb5os       & 0.11868356\\\\\n",
       "\t2 & Tfap2a & Gm26688       & 0.11001707\\\\\n",
       "\t3 & Rfx6   & Hnf1b         & 0.09558139\\\\\n",
       "\t4 & Sim1   & Otx2os1       & 0.09500712\\\\\n",
       "\t5 & Tfap2a & Tnfrsf11b     & 0.09362741\\\\\n",
       "\t6 & Celf4  & Anxa1         & 0.09282246\\\\\n",
       "\t7 & Tfap2a & Tfap2b        & 0.09205630\\\\\n",
       "\t8 & Terf1  & Pif1          & 0.09076750\\\\\n",
       "\t9 & Esrrg  & Stac2         & 0.09070625\\\\\n",
       "\t10 & Tfap2b & Wnt3          & 0.09022312\\\\\n",
       "\t11 & Terf1  & Lockd         & 0.08919815\\\\\n",
       "\t12 & Terf1  & Cdkn3         & 0.08498176\\\\\n",
       "\t13 & Otx2   & Otx2os1       & 0.08487994\\\\\n",
       "\t14 & Rfx6   & Enpp1         & 0.08339922\\\\\n",
       "\t15 & Esrrg  & Fam20a        & 0.08308780\\\\\n",
       "\t16 & Sim1   & Otx2          & 0.08301071\\\\\n",
       "\t17 & Prrx1  & Kcnk10        & 0.08176160\\\\\n",
       "\t18 & Esrrg  & Casr          & 0.08147499\\\\\n",
       "\t19 & Tfap2b & Cxcl14        & 0.08043888\\\\\n",
       "\t20 & Tfap2a & Wnt3          & 0.08020572\\\\\n",
       "\t21 & Esrrg  & Tsc22d3       & 0.07962714\\\\\n",
       "\t22 & Pou6f2 & Vwc2          & 0.07940138\\\\\n",
       "\t23 & Hnf1b  & Rfx6          & 0.07920826\\\\\n",
       "\t24 & Otx2   & Postn         & 0.07892602\\\\\n",
       "\t25 & Terf1  & Cenpf         & 0.07885115\\\\\n",
       "\t26 & Terf1  & Ccnb1         & 0.07878102\\\\\n",
       "\t27 & Tfap2a & Hgf           & 0.07871124\\\\\n",
       "\t28 & Tfap2b & Tnfrsf11b     & 0.07800308\\\\\n",
       "\t29 & Tfap2b & E330013P04Rik & 0.07777844\\\\\n",
       "\t30 & Tfap2b & Tfap2a        & 0.07746201\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t116594 & Bptf    & Zbtb44        & 0.01000139\\\\\n",
       "\t116595 & Esrp1   & Usp20         & 0.01000130\\\\\n",
       "\t116596 & Tcf12   & Unc13b        & 0.01000119\\\\\n",
       "\t116597 & Thoc2   & Rcor1         & 0.01000105\\\\\n",
       "\t116598 & Larp1   & Rcc2          & 0.01000105\\\\\n",
       "\t116599 & Stat5b  & Col22a1       & 0.01000103\\\\\n",
       "\t116600 & Nfatc3  & Opa1          & 0.01000102\\\\\n",
       "\t116601 & Tfcp2   & Pou6f1        & 0.01000098\\\\\n",
       "\t116602 & Tcf12   & Rabep1        & 0.01000098\\\\\n",
       "\t116603 & Fubp1   & Hn1           & 0.01000096\\\\\n",
       "\t116604 & Fubp1   & Rcbtb2        & 0.01000093\\\\\n",
       "\t116605 & Cux1    & Slc37a3       & 0.01000086\\\\\n",
       "\t116606 & Rbms1   & Pias1         & 0.01000085\\\\\n",
       "\t116607 & Ctbp2   & Atrn          & 0.01000082\\\\\n",
       "\t116608 & Parp1   & Glod4         & 0.01000079\\\\\n",
       "\t116609 & Zfp148  & Ccdc88a       & 0.01000079\\\\\n",
       "\t116610 & Nfatc3  & Cpne3         & 0.01000077\\\\\n",
       "\t116611 & Zfp652  & Pkp4          & 0.01000068\\\\\n",
       "\t116612 & Mapk1   & Rab6a         & 0.01000067\\\\\n",
       "\t116613 & Rbfox2  & Heatr5b       & 0.01000066\\\\\n",
       "\t116614 & Tbl1xr1 & Oxct1         & 0.01000063\\\\\n",
       "\t116615 & Asap3   & Snx7          & 0.01000054\\\\\n",
       "\t116616 & Suz12   & Nudcd2        & 0.01000047\\\\\n",
       "\t116617 & Klf12   & Lin7a         & 0.01000046\\\\\n",
       "\t116618 & Xpa     & Cacna1a       & 0.01000030\\\\\n",
       "\t116619 & Rab2a   & Rai1          & 0.01000027\\\\\n",
       "\t116620 & Zfp521  & Capsl         & 0.01000020\\\\\n",
       "\t116621 & Lcor    & Pip4k2c       & 0.01000007\\\\\n",
       "\t116622 & Ikzf2   & 4930480K23Rik & 0.01000004\\\\\n",
       "\t116623 & Nrf1    & Paip2b        & 0.01000001\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 116623 × 3\n",
       "\n",
       "| <!--/--> | TF &lt;fct&gt; | Target &lt;fct&gt; | weight &lt;dbl&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | Hoxb6  | Hoxb5os       | 0.11868356 |\n",
       "| 2 | Tfap2a | Gm26688       | 0.11001707 |\n",
       "| 3 | Rfx6   | Hnf1b         | 0.09558139 |\n",
       "| 4 | Sim1   | Otx2os1       | 0.09500712 |\n",
       "| 5 | Tfap2a | Tnfrsf11b     | 0.09362741 |\n",
       "| 6 | Celf4  | Anxa1         | 0.09282246 |\n",
       "| 7 | Tfap2a | Tfap2b        | 0.09205630 |\n",
       "| 8 | Terf1  | Pif1          | 0.09076750 |\n",
       "| 9 | Esrrg  | Stac2         | 0.09070625 |\n",
       "| 10 | Tfap2b | Wnt3          | 0.09022312 |\n",
       "| 11 | Terf1  | Lockd         | 0.08919815 |\n",
       "| 12 | Terf1  | Cdkn3         | 0.08498176 |\n",
       "| 13 | Otx2   | Otx2os1       | 0.08487994 |\n",
       "| 14 | Rfx6   | Enpp1         | 0.08339922 |\n",
       "| 15 | Esrrg  | Fam20a        | 0.08308780 |\n",
       "| 16 | Sim1   | Otx2          | 0.08301071 |\n",
       "| 17 | Prrx1  | Kcnk10        | 0.08176160 |\n",
       "| 18 | Esrrg  | Casr          | 0.08147499 |\n",
       "| 19 | Tfap2b | Cxcl14        | 0.08043888 |\n",
       "| 20 | Tfap2a | Wnt3          | 0.08020572 |\n",
       "| 21 | Esrrg  | Tsc22d3       | 0.07962714 |\n",
       "| 22 | Pou6f2 | Vwc2          | 0.07940138 |\n",
       "| 23 | Hnf1b  | Rfx6          | 0.07920826 |\n",
       "| 24 | Otx2   | Postn         | 0.07892602 |\n",
       "| 25 | Terf1  | Cenpf         | 0.07885115 |\n",
       "| 26 | Terf1  | Ccnb1         | 0.07878102 |\n",
       "| 27 | Tfap2a | Hgf           | 0.07871124 |\n",
       "| 28 | Tfap2b | Tnfrsf11b     | 0.07800308 |\n",
       "| 29 | Tfap2b | E330013P04Rik | 0.07777844 |\n",
       "| 30 | Tfap2b | Tfap2a        | 0.07746201 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 116594 | Bptf    | Zbtb44        | 0.01000139 |\n",
       "| 116595 | Esrp1   | Usp20         | 0.01000130 |\n",
       "| 116596 | Tcf12   | Unc13b        | 0.01000119 |\n",
       "| 116597 | Thoc2   | Rcor1         | 0.01000105 |\n",
       "| 116598 | Larp1   | Rcc2          | 0.01000105 |\n",
       "| 116599 | Stat5b  | Col22a1       | 0.01000103 |\n",
       "| 116600 | Nfatc3  | Opa1          | 0.01000102 |\n",
       "| 116601 | Tfcp2   | Pou6f1        | 0.01000098 |\n",
       "| 116602 | Tcf12   | Rabep1        | 0.01000098 |\n",
       "| 116603 | Fubp1   | Hn1           | 0.01000096 |\n",
       "| 116604 | Fubp1   | Rcbtb2        | 0.01000093 |\n",
       "| 116605 | Cux1    | Slc37a3       | 0.01000086 |\n",
       "| 116606 | Rbms1   | Pias1         | 0.01000085 |\n",
       "| 116607 | Ctbp2   | Atrn          | 0.01000082 |\n",
       "| 116608 | Parp1   | Glod4         | 0.01000079 |\n",
       "| 116609 | Zfp148  | Ccdc88a       | 0.01000079 |\n",
       "| 116610 | Nfatc3  | Cpne3         | 0.01000077 |\n",
       "| 116611 | Zfp652  | Pkp4          | 0.01000068 |\n",
       "| 116612 | Mapk1   | Rab6a         | 0.01000067 |\n",
       "| 116613 | Rbfox2  | Heatr5b       | 0.01000066 |\n",
       "| 116614 | Tbl1xr1 | Oxct1         | 0.01000063 |\n",
       "| 116615 | Asap3   | Snx7          | 0.01000054 |\n",
       "| 116616 | Suz12   | Nudcd2        | 0.01000047 |\n",
       "| 116617 | Klf12   | Lin7a         | 0.01000046 |\n",
       "| 116618 | Xpa     | Cacna1a       | 0.01000030 |\n",
       "| 116619 | Rab2a   | Rai1          | 0.01000027 |\n",
       "| 116620 | Zfp521  | Capsl         | 0.01000020 |\n",
       "| 116621 | Lcor    | Pip4k2c       | 0.01000007 |\n",
       "| 116622 | Ikzf2   | 4930480K23Rik | 0.01000004 |\n",
       "| 116623 | Nrf1    | Paip2b        | 0.01000001 |\n",
       "\n"
      ],
      "text/plain": [
       "       TF      Target        weight    \n",
       "1      Hoxb6   Hoxb5os       0.11868356\n",
       "2      Tfap2a  Gm26688       0.11001707\n",
       "3      Rfx6    Hnf1b         0.09558139\n",
       "4      Sim1    Otx2os1       0.09500712\n",
       "5      Tfap2a  Tnfrsf11b     0.09362741\n",
       "6      Celf4   Anxa1         0.09282246\n",
       "7      Tfap2a  Tfap2b        0.09205630\n",
       "8      Terf1   Pif1          0.09076750\n",
       "9      Esrrg   Stac2         0.09070625\n",
       "10     Tfap2b  Wnt3          0.09022312\n",
       "11     Terf1   Lockd         0.08919815\n",
       "12     Terf1   Cdkn3         0.08498176\n",
       "13     Otx2    Otx2os1       0.08487994\n",
       "14     Rfx6    Enpp1         0.08339922\n",
       "15     Esrrg   Fam20a        0.08308780\n",
       "16     Sim1    Otx2          0.08301071\n",
       "17     Prrx1   Kcnk10        0.08176160\n",
       "18     Esrrg   Casr          0.08147499\n",
       "19     Tfap2b  Cxcl14        0.08043888\n",
       "20     Tfap2a  Wnt3          0.08020572\n",
       "21     Esrrg   Tsc22d3       0.07962714\n",
       "22     Pou6f2  Vwc2          0.07940138\n",
       "23     Hnf1b   Rfx6          0.07920826\n",
       "24     Otx2    Postn         0.07892602\n",
       "25     Terf1   Cenpf         0.07885115\n",
       "26     Terf1   Ccnb1         0.07878102\n",
       "27     Tfap2a  Hgf           0.07871124\n",
       "28     Tfap2b  Tnfrsf11b     0.07800308\n",
       "29     Tfap2b  E330013P04Rik 0.07777844\n",
       "30     Tfap2b  Tfap2a        0.07746201\n",
       "⋮      ⋮       ⋮             ⋮         \n",
       "116594 Bptf    Zbtb44        0.01000139\n",
       "116595 Esrp1   Usp20         0.01000130\n",
       "116596 Tcf12   Unc13b        0.01000119\n",
       "116597 Thoc2   Rcor1         0.01000105\n",
       "116598 Larp1   Rcc2          0.01000105\n",
       "116599 Stat5b  Col22a1       0.01000103\n",
       "116600 Nfatc3  Opa1          0.01000102\n",
       "116601 Tfcp2   Pou6f1        0.01000098\n",
       "116602 Tcf12   Rabep1        0.01000098\n",
       "116603 Fubp1   Hn1           0.01000096\n",
       "116604 Fubp1   Rcbtb2        0.01000093\n",
       "116605 Cux1    Slc37a3       0.01000086\n",
       "116606 Rbms1   Pias1         0.01000085\n",
       "116607 Ctbp2   Atrn          0.01000082\n",
       "116608 Parp1   Glod4         0.01000079\n",
       "116609 Zfp148  Ccdc88a       0.01000079\n",
       "116610 Nfatc3  Cpne3         0.01000077\n",
       "116611 Zfp652  Pkp4          0.01000068\n",
       "116612 Mapk1   Rab6a         0.01000067\n",
       "116613 Rbfox2  Heatr5b       0.01000066\n",
       "116614 Tbl1xr1 Oxct1         0.01000063\n",
       "116615 Asap3   Snx7          0.01000054\n",
       "116616 Suz12   Nudcd2        0.01000047\n",
       "116617 Klf12   Lin7a         0.01000046\n",
       "116618 Xpa     Cacna1a       0.01000030\n",
       "116619 Rab2a   Rai1          0.01000027\n",
       "116620 Zfp521  Capsl         0.01000020\n",
       "116621 Lcor    Pip4k2c       0.01000007\n",
       "116622 Ikzf2   4930480K23Rik 0.01000004\n",
       "116623 Nrf1    Paip2b        0.01000001"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scenic_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "1049a94a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Load a given gene regulatory (sub-)network.\n",
    "#' \n",
    "#' @param grn_name @param dge_name This function looks in here(\"networks\", grn_name, \"networks\", subnetwork_name).\n",
    "#' \n",
    "load_grn_by_subnetwork = function( grn_name, subnetwork_name, sort_by_weight = T, ... ){\n",
    "  grn_location = here(\"networks\", grn_name, \"networks\", subnetwork_name)\n",
    "  if(is.null(grn_location)){\n",
    "    stop(\"Error locating grn! Please report this error to Eric.\\n\")\n",
    "  }\n",
    "  extension = tools::file_ext(gsub(\".gz\", \"\", grn_location))\n",
    "    print(grn_location)\n",
    "  if (extension %in% c(\"txt\", \"tsv\")){\n",
    "    X = read.table(grn_location, header = F, row.names = NULL, stringsAsFactors = F, ... )\n",
    "  } else if (extension==\"csv\"){\n",
    "    X = read.csv  (grn_location, header = F, row.names = NULL, stringsAsFactors = F, ... ) \n",
    "  } else {\n",
    "    stop(\"Unknown format!\")\n",
    "  }\n",
    "  \n",
    "  # add score of -1 if missing\n",
    "  if(ncol(X) == 2){\n",
    "    X[[3]] = -1\n",
    "  }\n",
    "  \n",
    "  # Fix meanings of columns\n",
    "  first_column_meaning = load_grn_metadata()[grn_name, \"first_column\"]\n",
    "  if(first_column_meaning %in% c(\"symmetric\", EXPECTED_GRN_COLNAMES[[1]])){\n",
    "    colnames(X) = EXPECTED_GRN_COLNAMES\n",
    "  } else if(first_column_meaning == EXPECTED_GRN_COLNAMES[[2]]){\n",
    "    colnames(X) = EXPECTED_GRN_COLNAMES[c(2, 1, 3)]\n",
    "  } else {\n",
    "    stop(\"Can't tell if target or regulator comes first! Fix the metadata.\\n\")\n",
    "  }\n",
    "  \n",
    "  if(sort_by_weight){\n",
    "    X = X[order(X[[EXPECTED_GRN_COLNAMES[[3]]]], decreasing = T),]\n",
    "  }\n",
    "  \n",
    "  return(X[EXPECTED_GRN_COLNAMES])\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "ec031c03",
   "metadata": {},
   "outputs": [],
   "source": [
    "load_geneset = function( geneset_name, uppercase = T,  ... ){\n",
    "    geneset_location = here(geneset_name )\n",
    "  if(is.null(geneset_location)){\n",
    "    stop(\"Error locating geneset! Please report this error to Eric.\\n\")\n",
    "  }\n",
    "  X = read.csv(geneset_location, header = F, row.names = NULL, ..., stringsAsFactors = F )[[1]]\n",
    "  if(uppercase){\n",
    "    X %<>% toupper\n",
    "  }\n",
    "  if(toupper(X[[1]] == \"GENE\")){\n",
    "    X = X[-1]\n",
    "  }\n",
    "  return(X)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "6df15da1",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "280be60d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/Users/LoboM/Dropbox (UMass Medical School)/sharedUMass_Macrina_Rene/rerun_scenic/scripts/networks/scenic_pharynx/networks/pharynx_atlas.csv\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3672138 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>regulator</th><th scope=col>target</th><th scope=col>weight</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Hoxb6 </td><td>Hoxb5os      </td><td>0.10228378</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>Tfap2a</td><td>Gm26688      </td><td>0.10142356</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>Esrrg </td><td>Gcm2         </td><td>0.09993458</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Tfap2a</td><td>Tnfrsf11b    </td><td>0.09987369</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>Tfap2b</td><td>Cxcl14       </td><td>0.09715990</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>Tfap2a</td><td>Wnt9b        </td><td>0.09483255</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>Celf4 </td><td>Anxa1        </td><td>0.09125103</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>Sim1  </td><td>Otx2os1      </td><td>0.08997949</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>Esrrg </td><td>Stac2        </td><td>0.08891674</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>Pax8  </td><td>Kcnj16       </td><td>0.08712814</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>Gli3  </td><td>Rarres1      </td><td>0.08622034</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>Rfx6  </td><td>Hnf1b        </td><td>0.08610888</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>Tfap2b</td><td>Wnt3         </td><td>0.08407824</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>Otx2  </td><td>Otx2os1      </td><td>0.08343711</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>Otx2  </td><td>Postn        </td><td>0.08314272</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>Tfap2a</td><td>Tfap2b       </td><td>0.08197262</td></tr>\n",
       "\t<tr><th scope=row>17</th><td>Tfap2a</td><td>Hgf          </td><td>0.08183546</td></tr>\n",
       "\t<tr><th scope=row>18</th><td>Esrrg </td><td>Casr         </td><td>0.07952289</td></tr>\n",
       "\t<tr><th scope=row>19</th><td>Rfx6  </td><td>Enpp1        </td><td>0.07945111</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>Terf1 </td><td>Cdkn3        </td><td>0.07899770</td></tr>\n",
       "\t<tr><th scope=row>21</th><td>Fli1  </td><td>Pde1a        </td><td>0.07897190</td></tr>\n",
       "\t<tr><th scope=row>22</th><td>Esrrg </td><td>Fam20a       </td><td>0.07849946</td></tr>\n",
       "\t<tr><th scope=row>23</th><td>Terf1 </td><td>Cenpf        </td><td>0.07813805</td></tr>\n",
       "\t<tr><th scope=row>24</th><td>Gcm2  </td><td>Stac2        </td><td>0.07795471</td></tr>\n",
       "\t<tr><th scope=row>25</th><td>Terf1 </td><td>Ccnb1        </td><td>0.07760936</td></tr>\n",
       "\t<tr><th scope=row>26</th><td>Rfx6  </td><td>Gata6        </td><td>0.07690102</td></tr>\n",
       "\t<tr><th scope=row>27</th><td>Rfx6  </td><td>2410017I17Rik</td><td>0.07682477</td></tr>\n",
       "\t<tr><th scope=row>28</th><td>Tfap2a</td><td>Wnt3         </td><td>0.07676816</td></tr>\n",
       "\t<tr><th scope=row>29</th><td>Rfx6  </td><td>Slc17a2      </td><td>0.07668695</td></tr>\n",
       "\t<tr><th scope=row>30</th><td>Terf1 </td><td>Arl6ip1      </td><td>0.07574987</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>3672109</th><td>Apex2  </td><td>Ushbp1        </td><td>0.001000006</td></tr>\n",
       "\t<tr><th scope=row>3672110</th><td>Arid3b </td><td>Eif2s3x       </td><td>0.001000006</td></tr>\n",
       "\t<tr><th scope=row>3672111</th><td>Zfp131 </td><td>Rdm1          </td><td>0.001000006</td></tr>\n",
       "\t<tr><th scope=row>3672112</th><td>Npas2  </td><td>Klf7          </td><td>0.001000006</td></tr>\n",
       "\t<tr><th scope=row>3672113</th><td>Smap2  </td><td>Prrx1         </td><td>0.001000005</td></tr>\n",
       "\t<tr><th scope=row>3672114</th><td>Zbtb7b </td><td>Mapre3        </td><td>0.001000005</td></tr>\n",
       "\t<tr><th scope=row>3672115</th><td>Prox1  </td><td>Phb2          </td><td>0.001000005</td></tr>\n",
       "\t<tr><th scope=row>3672116</th><td>Smad6  </td><td>Hkdc1         </td><td>0.001000004</td></tr>\n",
       "\t<tr><th scope=row>3672117</th><td>Klf6   </td><td>Bcl2l14       </td><td>0.001000004</td></tr>\n",
       "\t<tr><th scope=row>3672118</th><td>Zfp384 </td><td>Nxpe5         </td><td>0.001000004</td></tr>\n",
       "\t<tr><th scope=row>3672119</th><td>Zfp110 </td><td>2610044O15Rik8</td><td>0.001000004</td></tr>\n",
       "\t<tr><th scope=row>3672120</th><td>Zfp62  </td><td>C130083M11Rik </td><td>0.001000004</td></tr>\n",
       "\t<tr><th scope=row>3672121</th><td>Mef2c  </td><td>Capns1        </td><td>0.001000004</td></tr>\n",
       "\t<tr><th scope=row>3672122</th><td>Pick1  </td><td>Pde2a         </td><td>0.001000004</td></tr>\n",
       "\t<tr><th scope=row>3672123</th><td>Zfp362 </td><td>Oc90          </td><td>0.001000003</td></tr>\n",
       "\t<tr><th scope=row>3672124</th><td>Hnrnph3</td><td>Hdac11        </td><td>0.001000003</td></tr>\n",
       "\t<tr><th scope=row>3672125</th><td>Sp4    </td><td>Ttc1          </td><td>0.001000003</td></tr>\n",
       "\t<tr><th scope=row>3672126</th><td>Runx2  </td><td>Klhdc8a       </td><td>0.001000003</td></tr>\n",
       "\t<tr><th scope=row>3672127</th><td>Setbp1 </td><td>Dcaf13        </td><td>0.001000003</td></tr>\n",
       "\t<tr><th scope=row>3672128</th><td>Ppara  </td><td>Zfp600        </td><td>0.001000002</td></tr>\n",
       "\t<tr><th scope=row>3672129</th><td>Mga    </td><td>Gm1976        </td><td>0.001000002</td></tr>\n",
       "\t<tr><th scope=row>3672130</th><td>Zbtb11 </td><td>Them4         </td><td>0.001000002</td></tr>\n",
       "\t<tr><th scope=row>3672131</th><td>E2f5   </td><td>Ackr4         </td><td>0.001000002</td></tr>\n",
       "\t<tr><th scope=row>3672132</th><td>Lrrfip1</td><td>Col6a2        </td><td>0.001000002</td></tr>\n",
       "\t<tr><th scope=row>3672133</th><td>Tead1  </td><td>Fam183b       </td><td>0.001000001</td></tr>\n",
       "\t<tr><th scope=row>3672134</th><td>Snd1   </td><td>Pik3r2        </td><td>0.001000001</td></tr>\n",
       "\t<tr><th scope=row>3672135</th><td>Rcor1  </td><td>Rfxap         </td><td>0.001000001</td></tr>\n",
       "\t<tr><th scope=row>3672136</th><td>Dab2   </td><td>2310043O21Rik </td><td>0.001000001</td></tr>\n",
       "\t<tr><th scope=row>3672137</th><td>Nkx2-1 </td><td>Arhgap20os    </td><td>0.001000001</td></tr>\n",
       "\t<tr><th scope=row>3672138</th><td>Zfp654 </td><td>AU019990      </td><td>0.001000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3672138 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & regulator & target & weight\\\\\n",
       "  & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & Hoxb6  & Hoxb5os       & 0.10228378\\\\\n",
       "\t2 & Tfap2a & Gm26688       & 0.10142356\\\\\n",
       "\t3 & Esrrg  & Gcm2          & 0.09993458\\\\\n",
       "\t4 & Tfap2a & Tnfrsf11b     & 0.09987369\\\\\n",
       "\t5 & Tfap2b & Cxcl14        & 0.09715990\\\\\n",
       "\t6 & Tfap2a & Wnt9b         & 0.09483255\\\\\n",
       "\t7 & Celf4  & Anxa1         & 0.09125103\\\\\n",
       "\t8 & Sim1   & Otx2os1       & 0.08997949\\\\\n",
       "\t9 & Esrrg  & Stac2         & 0.08891674\\\\\n",
       "\t10 & Pax8   & Kcnj16        & 0.08712814\\\\\n",
       "\t11 & Gli3   & Rarres1       & 0.08622034\\\\\n",
       "\t12 & Rfx6   & Hnf1b         & 0.08610888\\\\\n",
       "\t13 & Tfap2b & Wnt3          & 0.08407824\\\\\n",
       "\t14 & Otx2   & Otx2os1       & 0.08343711\\\\\n",
       "\t15 & Otx2   & Postn         & 0.08314272\\\\\n",
       "\t16 & Tfap2a & Tfap2b        & 0.08197262\\\\\n",
       "\t17 & Tfap2a & Hgf           & 0.08183546\\\\\n",
       "\t18 & Esrrg  & Casr          & 0.07952289\\\\\n",
       "\t19 & Rfx6   & Enpp1         & 0.07945111\\\\\n",
       "\t20 & Terf1  & Cdkn3         & 0.07899770\\\\\n",
       "\t21 & Fli1   & Pde1a         & 0.07897190\\\\\n",
       "\t22 & Esrrg  & Fam20a        & 0.07849946\\\\\n",
       "\t23 & Terf1  & Cenpf         & 0.07813805\\\\\n",
       "\t24 & Gcm2   & Stac2         & 0.07795471\\\\\n",
       "\t25 & Terf1  & Ccnb1         & 0.07760936\\\\\n",
       "\t26 & Rfx6   & Gata6         & 0.07690102\\\\\n",
       "\t27 & Rfx6   & 2410017I17Rik & 0.07682477\\\\\n",
       "\t28 & Tfap2a & Wnt3          & 0.07676816\\\\\n",
       "\t29 & Rfx6   & Slc17a2       & 0.07668695\\\\\n",
       "\t30 & Terf1  & Arl6ip1       & 0.07574987\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t3672109 & Apex2   & Ushbp1         & 0.001000006\\\\\n",
       "\t3672110 & Arid3b  & Eif2s3x        & 0.001000006\\\\\n",
       "\t3672111 & Zfp131  & Rdm1           & 0.001000006\\\\\n",
       "\t3672112 & Npas2   & Klf7           & 0.001000006\\\\\n",
       "\t3672113 & Smap2   & Prrx1          & 0.001000005\\\\\n",
       "\t3672114 & Zbtb7b  & Mapre3         & 0.001000005\\\\\n",
       "\t3672115 & Prox1   & Phb2           & 0.001000005\\\\\n",
       "\t3672116 & Smad6   & Hkdc1          & 0.001000004\\\\\n",
       "\t3672117 & Klf6    & Bcl2l14        & 0.001000004\\\\\n",
       "\t3672118 & Zfp384  & Nxpe5          & 0.001000004\\\\\n",
       "\t3672119 & Zfp110  & 2610044O15Rik8 & 0.001000004\\\\\n",
       "\t3672120 & Zfp62   & C130083M11Rik  & 0.001000004\\\\\n",
       "\t3672121 & Mef2c   & Capns1         & 0.001000004\\\\\n",
       "\t3672122 & Pick1   & Pde2a          & 0.001000004\\\\\n",
       "\t3672123 & Zfp362  & Oc90           & 0.001000003\\\\\n",
       "\t3672124 & Hnrnph3 & Hdac11         & 0.001000003\\\\\n",
       "\t3672125 & Sp4     & Ttc1           & 0.001000003\\\\\n",
       "\t3672126 & Runx2   & Klhdc8a        & 0.001000003\\\\\n",
       "\t3672127 & Setbp1  & Dcaf13         & 0.001000003\\\\\n",
       "\t3672128 & Ppara   & Zfp600         & 0.001000002\\\\\n",
       "\t3672129 & Mga     & Gm1976         & 0.001000002\\\\\n",
       "\t3672130 & Zbtb11  & Them4          & 0.001000002\\\\\n",
       "\t3672131 & E2f5    & Ackr4          & 0.001000002\\\\\n",
       "\t3672132 & Lrrfip1 & Col6a2         & 0.001000002\\\\\n",
       "\t3672133 & Tead1   & Fam183b        & 0.001000001\\\\\n",
       "\t3672134 & Snd1    & Pik3r2         & 0.001000001\\\\\n",
       "\t3672135 & Rcor1   & Rfxap          & 0.001000001\\\\\n",
       "\t3672136 & Dab2    & 2310043O21Rik  & 0.001000001\\\\\n",
       "\t3672137 & Nkx2-1  & Arhgap20os     & 0.001000001\\\\\n",
       "\t3672138 & Zfp654  & AU019990       & 0.001000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3672138 × 3\n",
       "\n",
       "| <!--/--> | regulator &lt;chr&gt; | target &lt;chr&gt; | weight &lt;dbl&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | Hoxb6  | Hoxb5os       | 0.10228378 |\n",
       "| 2 | Tfap2a | Gm26688       | 0.10142356 |\n",
       "| 3 | Esrrg  | Gcm2          | 0.09993458 |\n",
       "| 4 | Tfap2a | Tnfrsf11b     | 0.09987369 |\n",
       "| 5 | Tfap2b | Cxcl14        | 0.09715990 |\n",
       "| 6 | Tfap2a | Wnt9b         | 0.09483255 |\n",
       "| 7 | Celf4  | Anxa1         | 0.09125103 |\n",
       "| 8 | Sim1   | Otx2os1       | 0.08997949 |\n",
       "| 9 | Esrrg  | Stac2         | 0.08891674 |\n",
       "| 10 | Pax8   | Kcnj16        | 0.08712814 |\n",
       "| 11 | Gli3   | Rarres1       | 0.08622034 |\n",
       "| 12 | Rfx6   | Hnf1b         | 0.08610888 |\n",
       "| 13 | Tfap2b | Wnt3          | 0.08407824 |\n",
       "| 14 | Otx2   | Otx2os1       | 0.08343711 |\n",
       "| 15 | Otx2   | Postn         | 0.08314272 |\n",
       "| 16 | Tfap2a | Tfap2b        | 0.08197262 |\n",
       "| 17 | Tfap2a | Hgf           | 0.08183546 |\n",
       "| 18 | Esrrg  | Casr          | 0.07952289 |\n",
       "| 19 | Rfx6   | Enpp1         | 0.07945111 |\n",
       "| 20 | Terf1  | Cdkn3         | 0.07899770 |\n",
       "| 21 | Fli1   | Pde1a         | 0.07897190 |\n",
       "| 22 | Esrrg  | Fam20a        | 0.07849946 |\n",
       "| 23 | Terf1  | Cenpf         | 0.07813805 |\n",
       "| 24 | Gcm2   | Stac2         | 0.07795471 |\n",
       "| 25 | Terf1  | Ccnb1         | 0.07760936 |\n",
       "| 26 | Rfx6   | Gata6         | 0.07690102 |\n",
       "| 27 | Rfx6   | 2410017I17Rik | 0.07682477 |\n",
       "| 28 | Tfap2a | Wnt3          | 0.07676816 |\n",
       "| 29 | Rfx6   | Slc17a2       | 0.07668695 |\n",
       "| 30 | Terf1  | Arl6ip1       | 0.07574987 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 3672109 | Apex2   | Ushbp1         | 0.001000006 |\n",
       "| 3672110 | Arid3b  | Eif2s3x        | 0.001000006 |\n",
       "| 3672111 | Zfp131  | Rdm1           | 0.001000006 |\n",
       "| 3672112 | Npas2   | Klf7           | 0.001000006 |\n",
       "| 3672113 | Smap2   | Prrx1          | 0.001000005 |\n",
       "| 3672114 | Zbtb7b  | Mapre3         | 0.001000005 |\n",
       "| 3672115 | Prox1   | Phb2           | 0.001000005 |\n",
       "| 3672116 | Smad6   | Hkdc1          | 0.001000004 |\n",
       "| 3672117 | Klf6    | Bcl2l14        | 0.001000004 |\n",
       "| 3672118 | Zfp384  | Nxpe5          | 0.001000004 |\n",
       "| 3672119 | Zfp110  | 2610044O15Rik8 | 0.001000004 |\n",
       "| 3672120 | Zfp62   | C130083M11Rik  | 0.001000004 |\n",
       "| 3672121 | Mef2c   | Capns1         | 0.001000004 |\n",
       "| 3672122 | Pick1   | Pde2a          | 0.001000004 |\n",
       "| 3672123 | Zfp362  | Oc90           | 0.001000003 |\n",
       "| 3672124 | Hnrnph3 | Hdac11         | 0.001000003 |\n",
       "| 3672125 | Sp4     | Ttc1           | 0.001000003 |\n",
       "| 3672126 | Runx2   | Klhdc8a        | 0.001000003 |\n",
       "| 3672127 | Setbp1  | Dcaf13         | 0.001000003 |\n",
       "| 3672128 | Ppara   | Zfp600         | 0.001000002 |\n",
       "| 3672129 | Mga     | Gm1976         | 0.001000002 |\n",
       "| 3672130 | Zbtb11  | Them4          | 0.001000002 |\n",
       "| 3672131 | E2f5    | Ackr4          | 0.001000002 |\n",
       "| 3672132 | Lrrfip1 | Col6a2         | 0.001000002 |\n",
       "| 3672133 | Tead1   | Fam183b        | 0.001000001 |\n",
       "| 3672134 | Snd1    | Pik3r2         | 0.001000001 |\n",
       "| 3672135 | Rcor1   | Rfxap          | 0.001000001 |\n",
       "| 3672136 | Dab2    | 2310043O21Rik  | 0.001000001 |\n",
       "| 3672137 | Nkx2-1  | Arhgap20os     | 0.001000001 |\n",
       "| 3672138 | Zfp654  | AU019990       | 0.001000000 |\n",
       "\n"
      ],
      "text/plain": [
       "        regulator target         weight     \n",
       "1       Hoxb6     Hoxb5os        0.10228378 \n",
       "2       Tfap2a    Gm26688        0.10142356 \n",
       "3       Esrrg     Gcm2           0.09993458 \n",
       "4       Tfap2a    Tnfrsf11b      0.09987369 \n",
       "5       Tfap2b    Cxcl14         0.09715990 \n",
       "6       Tfap2a    Wnt9b          0.09483255 \n",
       "7       Celf4     Anxa1          0.09125103 \n",
       "8       Sim1      Otx2os1        0.08997949 \n",
       "9       Esrrg     Stac2          0.08891674 \n",
       "10      Pax8      Kcnj16         0.08712814 \n",
       "11      Gli3      Rarres1        0.08622034 \n",
       "12      Rfx6      Hnf1b          0.08610888 \n",
       "13      Tfap2b    Wnt3           0.08407824 \n",
       "14      Otx2      Otx2os1        0.08343711 \n",
       "15      Otx2      Postn          0.08314272 \n",
       "16      Tfap2a    Tfap2b         0.08197262 \n",
       "17      Tfap2a    Hgf            0.08183546 \n",
       "18      Esrrg     Casr           0.07952289 \n",
       "19      Rfx6      Enpp1          0.07945111 \n",
       "20      Terf1     Cdkn3          0.07899770 \n",
       "21      Fli1      Pde1a          0.07897190 \n",
       "22      Esrrg     Fam20a         0.07849946 \n",
       "23      Terf1     Cenpf          0.07813805 \n",
       "24      Gcm2      Stac2          0.07795471 \n",
       "25      Terf1     Ccnb1          0.07760936 \n",
       "26      Rfx6      Gata6          0.07690102 \n",
       "27      Rfx6      2410017I17Rik  0.07682477 \n",
       "28      Tfap2a    Wnt3           0.07676816 \n",
       "29      Rfx6      Slc17a2        0.07668695 \n",
       "30      Terf1     Arl6ip1        0.07574987 \n",
       "⋮       ⋮         ⋮              ⋮          \n",
       "3672109 Apex2     Ushbp1         0.001000006\n",
       "3672110 Arid3b    Eif2s3x        0.001000006\n",
       "3672111 Zfp131    Rdm1           0.001000006\n",
       "3672112 Npas2     Klf7           0.001000006\n",
       "3672113 Smap2     Prrx1          0.001000005\n",
       "3672114 Zbtb7b    Mapre3         0.001000005\n",
       "3672115 Prox1     Phb2           0.001000005\n",
       "3672116 Smad6     Hkdc1          0.001000004\n",
       "3672117 Klf6      Bcl2l14        0.001000004\n",
       "3672118 Zfp384    Nxpe5          0.001000004\n",
       "3672119 Zfp110    2610044O15Rik8 0.001000004\n",
       "3672120 Zfp62     C130083M11Rik  0.001000004\n",
       "3672121 Mef2c     Capns1         0.001000004\n",
       "3672122 Pick1     Pde2a          0.001000004\n",
       "3672123 Zfp362    Oc90           0.001000003\n",
       "3672124 Hnrnph3   Hdac11         0.001000003\n",
       "3672125 Sp4       Ttc1           0.001000003\n",
       "3672126 Runx2     Klhdc8a        0.001000003\n",
       "3672127 Setbp1    Dcaf13         0.001000003\n",
       "3672128 Ppara     Zfp600         0.001000002\n",
       "3672129 Mga       Gm1976         0.001000002\n",
       "3672130 Zbtb11    Them4          0.001000002\n",
       "3672131 E2f5      Ackr4          0.001000002\n",
       "3672132 Lrrfip1   Col6a2         0.001000002\n",
       "3672133 Tead1     Fam183b        0.001000001\n",
       "3672134 Snd1      Pik3r2         0.001000001\n",
       "3672135 Rcor1     Rfxap          0.001000001\n",
       "3672136 Dab2      2310043O21Rik  0.001000001\n",
       "3672137 Nkx2-1    Arhgap20os     0.001000001\n",
       "3672138 Zfp654    AU019990       0.001000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tf_target = load_grn_by_subnetwork(\"scenic_pharynx\", subnetwork_name = \"pharynx_atlas\" %>% paste0(\".csv\"))\n",
    "  tf_target_eval_subset = tf_target %>% subset(regulator == \"Foxn1\")\n",
    "  tf_target_eval_subset$in_zuklys = tf_target_eval_subset$target %in% load_geneset(\"foxn1_targets.csv\", uppercase = F)\n",
    "  p = ggplot(tf_target_eval_subset) + \n",
    "    geom_violin(aes(x = in_zuklys, fill = in_zuklys, y = weight)) +\n",
    "    geom_hline(  yintercept = 0.01 ) + \n",
    "    xlab(\"Confirmed vs unconfirmed Foxn1 targets\") + \n",
    "    ylab(\"weight\") + \n",
    "    ggtitle(\"Edge filtering threshold\")\n",
    "  ggsave(here(\"networks\", \"scenic_pharynx\", \"filtering.pdf\"), p, width = 4, height = 4)\n",
    "  \n",
    "  return(tf_target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef50a774",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
